

<!doctype html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>genipe.formats.impute2 &#8212; genipe 1.5.0 documentation</title>
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/bizstyle.css" type="text/css" />
    
    <script id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script src="../../../_static/bizstyle.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0" />
    <!--[if lt IE 9]>
    <script src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">genipe 1.5.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../../genipe.html" accesskey="U">genipe</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">genipe.formats.impute2</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for genipe.formats.impute2</h1><div class="highlight"><pre>
<span></span>
<span class="c1"># This file is part of genipe.</span>
<span class="c1">#</span>
<span class="c1"># This work is licensed under the Creative Commons Attribution-NonCommercial</span>
<span class="c1"># 4.0 International License. To view a copy of this license, visit</span>
<span class="c1"># http://creativecommons.org/licenses/by-nc/4.0/ or send a letter to Creative</span>
<span class="c1"># Commons, PO Box 1866, Mountain View, CA 94042, USA.</span>


<span class="kn">import</span> <span class="nn">logging</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">..error</span> <span class="kn">import</span> <span class="n">GenipeError</span>


<span class="n">__author__</span> <span class="o">=</span> <span class="s2">&quot;Louis-Philippe Lemieux Perreault&quot;</span>
<span class="n">__copyright__</span> <span class="o">=</span> <span class="s2">&quot;Copyright 2014, Beaulieu-Saucier Pharmacogenomics Centre&quot;</span>
<span class="n">__license__</span> <span class="o">=</span> <span class="s2">&quot;Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)&quot;</span>


<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;matrix_from_line&quot;</span><span class="p">,</span> <span class="s2">&quot;get_good_probs&quot;</span><span class="p">,</span> <span class="s2">&quot;maf_from_probs&quot;</span><span class="p">,</span>
           <span class="s2">&quot;dosage_from_probs&quot;</span><span class="p">,</span> <span class="s2">&quot;hard_calls_from_probs&quot;</span><span class="p">,</span>
           <span class="s2">&quot;maf_dosage_from_probs&quot;</span><span class="p">,</span> <span class="s2">&quot;additive_from_probs&quot;</span><span class="p">]</span>


<div class="viewcode-block" id="matrix_from_line"><a class="viewcode-back" href="../../../module_content/genipe.formats.html#genipe.formats.impute2.matrix_from_line">[docs]</a><span class="k">def</span> <span class="nf">matrix_from_line</span><span class="p">(</span><span class="n">impute2_line</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generates the probability matrix from an IMPUTE2 line.</span>

<span class="sd">    Args:</span>
<span class="sd">        impute2_line (list): a single line from IMPUTE2&#39;s result (split by</span>
<span class="sd">                             space)</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: a tuple containing the marker&#39;s information (first five values</span>
<span class="sd">               of the line) and the matrix probability (numpy array, float)</span>

<span class="sd">    The shape of the matrix is n x 3 where n is the number of samples.</span>
<span class="sd">    The columns represent the probability for AA, AB and BB.</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">        The ``impute2_line`` variable is a list of str, corresponding to a line</span>
<span class="sd">        from the IMPUTE2&#39;s result, split by space.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Creating the array and changing it&#39;s shape</span>
    <span class="n">probabilities</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">impute2_line</span><span class="p">[</span><span class="mi">5</span><span class="p">:],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">probabilities</span><span class="o">.</span><span class="n">shape</span> <span class="o">=</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">probabilities</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">impute2_line</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">probabilities</span></div>


<div class="viewcode-block" id="get_good_probs"><a class="viewcode-back" href="../../../module_content/genipe.formats.html#genipe.formats.impute2.get_good_probs">[docs]</a><span class="k">def</span> <span class="nf">get_good_probs</span><span class="p">(</span><span class="n">prob_matrix</span><span class="p">,</span> <span class="n">min_prob</span><span class="o">=</span><span class="mf">0.9</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Gathers good imputed genotypes (&gt;= probability threshold).</span>

<span class="sd">    Args:</span>
<span class="sd">        prob_matrix (numpy.array): the probability matrix</span>
<span class="sd">        min_prob (float): the probability threshold</span>

<span class="sd">    Returns:</span>
<span class="sd">        numpy.array: a mask array containing the positions where the</span>
<span class="sd">                     probabilities are equal or higher to the threshold</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">amax</span><span class="p">(</span><span class="n">prob_matrix</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">min_prob</span></div>


<div class="viewcode-block" id="maf_from_probs"><a class="viewcode-back" href="../../../module_content/genipe.formats.html#genipe.formats.impute2.maf_from_probs">[docs]</a><span class="k">def</span> <span class="nf">maf_from_probs</span><span class="p">(</span><span class="n">prob_matrix</span><span class="p">,</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">gender</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">site_name</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Computes MAF from a probability matrix (and gender if chromosome X).</span>

<span class="sd">    Args:</span>
<span class="sd">        prob_matrix (numpy.array): the probability matrix</span>
<span class="sd">        a1 (str): the first allele</span>
<span class="sd">        a2 (str): the second allele</span>
<span class="sd">        gender (numpy.array): the gender of the samples</span>
<span class="sd">        site_name (str): the name for this site</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: a tuple containing three values: the minor allele frequency, the</span>
<span class="sd">               minor and the major allele.</span>

<span class="sd">    When &#39;gender&#39; is not None, we assume that the MAF on chromosome X is</span>
<span class="sd">    required (hence, males count as 1, and females as 2 alleles). There is also</span>
<span class="sd">    an Exception raised if there are any heterozygous males.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># By default, the MAF is NA, and a1=major, a2=minor</span>
    <span class="n">maf</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
    <span class="n">major</span><span class="p">,</span> <span class="n">minor</span> <span class="o">=</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span>

    <span class="c1"># If there are no data, we return default values</span>
    <span class="k">if</span> <span class="n">prob_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">maf</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span>

    <span class="k">if</span> <span class="n">gender</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Not checking gender (this isn&#39;t chromosome X)</span>
        <span class="n">nb_geno</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">prob_matrix</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">minlength</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">maf</span> <span class="o">=</span> <span class="p">((</span><span class="n">nb_geno</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span> <span class="n">nb_geno</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">nb_geno</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Getting the males and females</span>
        <span class="n">males</span> <span class="o">=</span> <span class="p">(</span><span class="n">gender</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">females</span> <span class="o">=</span> <span class="p">(</span><span class="n">gender</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)</span>

        <span class="c1"># Male counts</span>
        <span class="n">males_nb_geno</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">prob_matrix</span><span class="p">[</span><span class="n">males</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
                                    <span class="n">minlength</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

        <span class="c1"># Female counts</span>
        <span class="n">females_nb_geno</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">prob_matrix</span><span class="p">[</span><span class="n">females</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
                                      <span class="n">minlength</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

        <span class="c1"># The total number of genotypes</span>
        <span class="n">total_geno_males</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">males_nb_geno</span><span class="p">)</span>
        <span class="n">total_geno_females</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">females_nb_geno</span><span class="p">)</span>

        <span class="c1"># If there are no genotypes, we return default values</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">total_geno_males</span> <span class="o">+</span> <span class="n">total_geno_females</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">maf</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span>

        <span class="c1"># There shouldn&#39;t be heterozygous genotypes for males</span>
        <span class="k">if</span> <span class="n">males_nb_geno</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: heterozygous male &quot;</span>
                              <span class="s2">&quot;present&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">site_name</span><span class="p">))</span>

        <span class="c1"># Computing the frequencies</span>
        <span class="n">maf</span> <span class="o">=</span> <span class="n">males_nb_geno</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="p">(</span><span class="n">females_nb_geno</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span> <span class="n">females_nb_geno</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">maf</span> <span class="o">/=</span> <span class="p">(</span><span class="n">total_geno_males</span> <span class="o">+</span> <span class="p">(</span><span class="n">total_geno_females</span> <span class="o">*</span> <span class="mi">2</span><span class="p">))</span>

    <span class="c1"># Is this the MAF?</span>
    <span class="k">if</span> <span class="n">maf</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">maf</span> <span class="o">&gt;</span> <span class="mf">0.5</span><span class="p">:</span>
        <span class="n">minor</span><span class="p">,</span> <span class="n">major</span> <span class="o">=</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span>
        <span class="n">maf</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">maf</span>

    <span class="k">return</span> <span class="n">maf</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span></div>


<div class="viewcode-block" id="maf_dosage_from_probs"><a class="viewcode-back" href="../../../module_content/genipe.formats.html#genipe.formats.impute2.maf_dosage_from_probs">[docs]</a><span class="k">def</span> <span class="nf">maf_dosage_from_probs</span><span class="p">(</span><span class="n">prob_matrix</span><span class="p">,</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">gender</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                          <span class="n">site_name</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Computes MAF and dosage vector from probs matrix.</span>

<span class="sd">    Args:</span>
<span class="sd">        prob_matrix (numpy.array): the probability matrix</span>
<span class="sd">        a1 (str): the first allele</span>
<span class="sd">        a2 (str): the second allele</span>
<span class="sd">        scale (int): the scale value</span>
<span class="sd">        gender (numpy.array): the gender of the samples</span>
<span class="sd">        site_name (str): the name for this site</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: a tuple containing four values: the dosage vector, the minor</span>
<span class="sd">               allele frequency, the minor and the major allele.</span>

<span class="sd">    When &#39;gender&#39; is not None, we assume that the MAF on chromosome X is</span>
<span class="sd">    required (hence, males count as 1, and females as 2 alleles). There is also</span>
<span class="sd">    an Exception raised if there are any heterozygous males.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># By default, the MAF is NA, and a1=major, a2=minor</span>
    <span class="n">maf</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
    <span class="n">major</span><span class="p">,</span> <span class="n">minor</span> <span class="o">=</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span>

    <span class="c1"># If there are no data, we return default values</span>
    <span class="k">if</span> <span class="n">prob_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">),</span> <span class="n">maf</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span>

    <span class="c1"># Getting the dosage by default (by default, on the second allele)</span>
    <span class="n">dosage</span> <span class="o">=</span> <span class="n">dosage_from_probs</span><span class="p">(</span>
        <span class="n">homo_probs</span><span class="o">=</span><span class="n">prob_matrix</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">],</span>
        <span class="n">hetero_probs</span><span class="o">=</span><span class="n">prob_matrix</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span>
        <span class="n">scale</span><span class="o">=</span><span class="n">scale</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">set_no_maf</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">if</span> <span class="n">gender</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Not checking gender (this isn&#39;t chromosome X)</span>
        <span class="n">maf</span> <span class="o">=</span> <span class="n">dosage</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dosage</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Getting the males and females</span>
        <span class="n">m</span> <span class="o">=</span> <span class="p">(</span><span class="n">gender</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">f</span> <span class="o">=</span> <span class="p">(</span><span class="n">gender</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)</span>

        <span class="c1"># Checking males genotype</span>
        <span class="n">males_nb_geno</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">prob_matrix</span><span class="p">[</span><span class="n">m</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
                                    <span class="n">minlength</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">males_nb_geno</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: heterozygous male &quot;</span>
                              <span class="s2">&quot;present&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">site_name</span><span class="p">))</span>

        <span class="c1"># The number of alleles</span>
        <span class="n">nb_alleles</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">+</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">nb_alleles</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># Gender is unknown for all, so we compute frequency for right</span>
            <span class="c1"># dosage value, then we set the MAF to NA afterwards</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;All samples have unknown gender, MAF will be NA&quot;</span><span class="p">)</span>
            <span class="n">maf</span> <span class="o">=</span> <span class="n">dosage</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dosage</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span>
            <span class="n">set_no_maf</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Computing frequencies</span>
            <span class="n">maf</span> <span class="o">=</span> <span class="p">(</span><span class="n">dosage</span><span class="p">[</span><span class="n">f</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">+</span> <span class="p">(</span><span class="n">dosage</span><span class="p">[</span><span class="n">m</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="mi">2</span><span class="p">))</span> <span class="o">/</span> <span class="n">nb_alleles</span>

    <span class="c1"># Is this the MAF?</span>
    <span class="k">if</span> <span class="n">maf</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">maf</span> <span class="o">&gt;</span> <span class="mf">0.5</span><span class="p">:</span>
        <span class="n">minor</span><span class="p">,</span> <span class="n">major</span> <span class="o">=</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span>
        <span class="n">maf</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">maf</span>
        <span class="n">dosage</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">-</span> <span class="n">dosage</span>

    <span class="k">return</span> <span class="n">dosage</span><span class="p">,</span> <span class="n">maf</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">set_no_maf</span> <span class="k">else</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span></div>


<div class="viewcode-block" id="dosage_from_probs"><a class="viewcode-back" href="../../../module_content/genipe.formats.html#genipe.formats.impute2.dosage_from_probs">[docs]</a><span class="k">def</span> <span class="nf">dosage_from_probs</span><span class="p">(</span><span class="n">homo_probs</span><span class="p">,</span> <span class="n">hetero_probs</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Computes dosage from probability matrix (for the minor allele).</span>

<span class="sd">    Args:</span>
<span class="sd">        homo_probs (numpy.array): the probabilities for the homozygous genotype</span>
<span class="sd">        hetero_probs (numpy.array): the probabilities for the heterozygous</span>
<span class="sd">                                    genotype</span>
<span class="sd">        scale (int): the scale value</span>

<span class="sd">    Returns:</span>
<span class="sd">        numpy.array: the dosage computed from the probabilities</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">homo_probs</span> <span class="o">+</span> <span class="p">(</span><span class="n">hetero_probs</span> <span class="o">/</span> <span class="mi">2</span><span class="p">))</span> <span class="o">*</span> <span class="n">scale</span></div>


<div class="viewcode-block" id="hard_calls_from_probs"><a class="viewcode-back" href="../../../module_content/genipe.formats.html#genipe.formats.impute2.hard_calls_from_probs">[docs]</a><span class="k">def</span> <span class="nf">hard_calls_from_probs</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">probs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Computes hard calls from probability matrix.</span>

<span class="sd">    Args:</span>
<span class="sd">        a1 (str): the first allele</span>
<span class="sd">        a2 (str): the second allele</span>
<span class="sd">        probs (numpy.array): the probability matrix</span>

<span class="sd">    Returns:</span>
<span class="sd">        numpy.array: the hard calls computed from the probabilities</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">possible_geno</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
        <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">a1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span><span class="p">),</span>     <span class="c1"># Homo A1</span>
        <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">]),</span>     <span class="c1"># Hetero</span>
        <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">a2</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span><span class="p">),</span>     <span class="c1"># Homo A2</span>
    <span class="p">])</span>

    <span class="k">return</span> <span class="n">possible_geno</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">probs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)]</span></div>


<div class="viewcode-block" id="additive_from_probs"><a class="viewcode-back" href="../../../module_content/genipe.formats.html#genipe.formats.impute2.additive_from_probs">[docs]</a><span class="k">def</span> <span class="nf">additive_from_probs</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">probs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Compute additive format from probability matrix.</span>

<span class="sd">    Args:</span>
<span class="sd">        a1 (str): the a1 allele</span>
<span class="sd">        a2 (str): the a2 allele</span>
<span class="sd">        probs (numpy.array): the probability matrix</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: the additive format computed from the probabilities, the minor</span>
<span class="sd">               and major allele.</span>

<span class="sd">    The encoding is as follow: 0 when homozygous major allele, 1 when</span>
<span class="sd">    heterozygous and 2 when homozygous minor allele.</span>

<span class="sd">    The minor and major alleles are inferred by looking at the MAF. By default,</span>
<span class="sd">    we think a2 is the minor allele, but flip if required.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">calls</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">probs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">minor</span> <span class="o">=</span> <span class="n">a2</span>
    <span class="n">major</span> <span class="o">=</span> <span class="n">a1</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">calls</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">calls</span><span class="p">)</span><span class="o">*</span><span class="mi">2</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mf">0.5</span><span class="p">:</span>
        <span class="n">calls</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">-</span> <span class="n">calls</span>
        <span class="n">minor</span> <span class="o">=</span> <span class="n">a1</span>
        <span class="n">major</span> <span class="o">=</span> <span class="n">a2</span>
    <span class="k">return</span> <span class="n">calls</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span></div>
</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">genipe 1.5.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../../genipe.html" >genipe</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">genipe.formats.impute2</a></li> 
      </ul>
    </div>


    <div class="footer" role="contentinfo">
        &#169; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 3.4.3.
    </div>

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-81009922-1', 'auto');
  ga('send', 'pageview');

</script>


  </body>
</html>