

<!doctype html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Site extraction &#8212; genipe 1.5.0 documentation</title>
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bizstyle.css" type="text/css" />
    
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script src="../_static/bizstyle.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Statistical Analysis Execution Time" href="../execution_time.html" />
    <link rel="prev" title="Logistic regression" href="tutorial_logistic.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0" />
    <!--[if lt IE 9]>
    <script src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../execution_time.html" title="Statistical Analysis Execution Time"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_logistic.html" title="Logistic regression"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">genipe 1.5.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorials.html" accesskey="U">Short Tutorials</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Site extraction</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="contents topic" id="quick-navigation">
<p class="topic-title">Quick navigation</p>
<ul class="simple">
<li><p><a class="reference internal" href="#site-extraction" id="id1">Site extraction</a></p>
<ul>
<li><p><a class="reference internal" href="#input-files" id="id2">Input files</a></p></li>
<li><p><a class="reference internal" href="#executing-the-extraction" id="id3">Executing the extraction</a></p></li>
<li><p><a class="reference internal" href="#output-files" id="id4">Output files</a></p></li>
<li><p><a class="reference internal" href="#usage" id="id5">Usage</a></p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="site-extraction">
<h1><a class="toc-backref" href="#id1">Site extraction</a><a class="headerlink" href="#site-extraction" title="Permalink to this headline">¶</a></h1>
<p>Genome-wide imputation dataset might be huge. Often, it is required to extract
a subset of imputed sites (<em>e.g.</em> specific markers, genomic location, or
markers with a specific minor allele frequency, information value or completion
rate). Also, different format might be required, depending of the underlying
analysis (<em>e.g.</em> hard calls or dosage values). We provide an easy tool to
perform site extraction of multiple <em>impute2</em> files using either marker
identification number, or genomic location and/or minor allele frequency and/or
call rate and/or information value.</p>
<p>We suppose that you have followed the main <a class="reference internal" href="tutorial_genipe.html#genipe-tut-page"><span class="std std-ref">Genome-wide imputation pipeline</span></a>. The
following command will create the working directory for this tutorial.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir -p <span class="nv">$HOME</span>/genipe_tutorial/extraction
</pre></div>
</div>
<div class="section" id="input-files">
<span id="extract-tut-input-files"></span><h2><a class="toc-backref" href="#id2">Input files</a><a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h2>
<p>After running the <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genipe</span></code></a> pipeline, all the required files for the
extraction tools are automatically created in the <code class="docutils literal notranslate"><span class="pre">final_impute2</span></code> directories
(see the <a class="reference internal" href="tutorial_genipe.html#genipe-tut-output-files-final-impute2"><span class="std std-ref">genipe/chrN/final_impute2 directories</span></a> section in the main
<a class="reference internal" href="tutorial_genipe.html#genipe-tut-page"><span class="std std-ref">Genome-wide imputation pipeline</span></a>).</p>
<p>The files that are required in these directories depends of what kind of
extraction is required (by name, or by genomic location and/or by minor allele
frequency and/or by calling rate and/or by information value).</p>
<p>Once the required <em>impute2</em> files are provided to the tool, the other required
files will be automatically fetched (if required).</p>
</div>
<div class="section" id="executing-the-extraction">
<span id="extract-tut-execute"></span><h2><a class="toc-backref" href="#id3">Executing the extraction</a><a class="headerlink" href="#executing-the-extraction" title="Permalink to this headline">¶</a></h2>
<p>The first time the tool is used on a set of <em>impute2</em> files, indexation will
automatically occur (to speed of the analysis for future extraction). There are
two ways to extract markers: using their identification number (<code class="docutils literal notranslate"><span class="pre">--extract</span></code>),
or using their properties (<code class="docutils literal notranslate"><span class="pre">--genomic</span></code>, <code class="docutils literal notranslate"><span class="pre">--maf</span></code>, <code class="docutils literal notranslate"><span class="pre">--rate</span></code> and/or
<code class="docutils literal notranslate"><span class="pre">--info</span></code>).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>It is possible to extract from multiple <em>impute2</em> files at the same time (by
specifying multiple input files).</p>
</div>
<div class="section" id="extraction-by-id">
<h3>Extraction by ID<a class="headerlink" href="#extraction-by-id" title="Permalink to this headline">¶</a></h3>
<p>To extract markers using their identification number, you need a file
containing the list of marker to extract (one marker per line).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial/extraction

<span class="nb">echo</span> <span class="s2">&quot;rs76139713:51137523:C:T&quot;</span> &gt; marker_list.txt
<span class="nb">echo</span> <span class="s2">&quot;rs372879164:17037188:A:G&quot;</span> &gt;&gt; marker_list.txt
</pre></div>
</div>
<p>This <code class="docutils literal notranslate"><span class="pre">marker_list.txt</span></code> file will contain the following:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>rs76139713:51137523:C:T
rs372879164:17037188:A:G
</pre></div>
</div>
<p>Then, the following command (using the <code class="docutils literal notranslate"><span class="pre">--extract</span></code> option) will extract those
two markers from the <em>impute2</em> file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>impute2-extractor <span class="se">\</span>
    --impute2 ../genipe/chr22/final_impute2/chr22.imputed.impute2.gz <span class="se">\</span>
    --extract marker_list.txt
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To gather a list of marker identification numbers, refer to the file
<code class="docutils literal notranslate"><span class="pre">chr22.imputed.map</span></code>, which contains the list of all sites in the <em>impute2</em>
file.</p>
</div>
</div>
<div class="section" id="extraction-by-characteristics">
<h3>Extraction by characteristics<a class="headerlink" href="#extraction-by-characteristics" title="Permalink to this headline">¶</a></h3>
<p>There are four ways to extract markers according to their characteristics. The
first way is to specify the genomic location of the markers to extract (<em>i.e.</em>
the <code class="docutils literal notranslate"><span class="pre">--genomic</span></code> option). The second way is to specify a minor allele
frequency threshold (<em>i.e.</em> the <code class="docutils literal notranslate"><span class="pre">--maf</span></code> option). The third way is to specify
a call rate threshold (<em>i.e.</em> the <code class="docutils literal notranslate"><span class="pre">--rate</span></code> option). The fourth and final way
is to specify an information value threshold (<em>i.e.</em> the <code class="docutils literal notranslate"><span class="pre">--info</span></code> option).
Those four ways can be used at the same time (<em>e.g.</em> to get markers in a
specific genomic range and a specific call rate).</p>
<p>For example, to extract markers with a MAF <span class="math notranslate nohighlight">\(\geq\)</span> 0.05 located in the
<em>CYP2D6</em> gene, perform the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial/extraction

impute2-extractor <span class="se">\</span>
    --impute2 ../genipe/chr22/final_impute2/chr22.imputed.impute2.gz <span class="se">\</span>
    --genomic chr22:42522501-42526883 <span class="se">\</span>
    --maf <span class="m">0</span>.05 <span class="se">\</span>
    --out cyp2d6_common
</pre></div>
</div>
<p>To gather all markers with a MAF <span class="math notranslate nohighlight">\(\geq\)</span> 0.05 and a call rate <span class="math notranslate nohighlight">\(\geq\)</span>
0.99, perform the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>impute2-extractor <span class="se">\</span>
    --impute2 ../genipe/chr22/final_impute2/chr22.imputed.impute2.gz <span class="se">\</span>
    --maf <span class="m">0</span>.05 <span class="se">\</span>
    --rate <span class="m">0</span>.99 <span class="se">\</span>
    --out common_complete
</pre></div>
</div>
</div>
</div>
<div class="section" id="output-files">
<span id="extract-tut-output-files"></span><h2><a class="toc-backref" href="#id4">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>The output files will depend on the output format selected (the <code class="docutils literal notranslate"><span class="pre">--format</span></code>
option). You can specify either <code class="docutils literal notranslate"><span class="pre">impute2</span></code>, <code class="docutils literal notranslate"><span class="pre">dosage</span></code>, <code class="docutils literal notranslate"><span class="pre">calls</span></code> and/or
<code class="docutils literal notranslate"><span class="pre">bed</span></code>, for the <em>impute2</em> format (<em>i.e.</em> three probabilities per sample), the
<em>dosage</em> format (<em>i.e.</em> one value between 0 and 2 per sample), hard calls and
binary <em>Plink</em> file.</p>
<div class="section" id="impute2-file">
<h3><code class="docutils literal notranslate"><span class="pre">.impute2</span></code> file<a class="headerlink" href="#impute2-file" title="Permalink to this headline">¶</a></h3>
<p>This file is generated when the <code class="docutils literal notranslate"><span class="pre">impute2</span></code> format is used. It has the same
format as the original <em>impute2</em> file.</p>
<p>The general structure of the file contains the following columns (which are
space delimited): the chromosome, the name of the marker, its position and its
two alleles. The subsequent columns correspond to the probabilities of each
genotype (hence, there are three columns per sample). The first value
correspond to the probability of being homozygous of the first allele. The
second value correspond to the probability of being heterozygous. Finally, the
third value correspond to the probability of being homozygous of the second
allele.</p>
<p>The following example shows two lines of the <code class="docutils literal notranslate"><span class="pre">.impute2</span></code> file.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>22 rs7289830 16058758 C A 0 0 1 0 0 1 0 1 0 ...
22 rs6423472 16087621 A G 0 1 0 1 0 0 0 1 0 ...
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When extracting using the <code class="docutils literal notranslate"><span class="pre">impute2</span></code> format, all the existing companion
files (<code class="docutils literal notranslate"><span class="pre">.maf</span></code>, <code class="docutils literal notranslate"><span class="pre">.map</span></code>, etc.) will also be extracted and included in the
same directory (using the same output prefix).</p>
</div>
</div>
<div class="section" id="dosage-file">
<h3><code class="docutils literal notranslate"><span class="pre">.dosage</span></code> file<a class="headerlink" href="#dosage-file" title="Permalink to this headline">¶</a></h3>
<p>This file contains the dosage computed from the <em>impute2</em> probabilities. The
general structure of the file contains the following columns (which are
tabulation separated): the chromosome, the position on the chromosome, its
name, its minor and major allele and the dosage value. The dosage values vary
between 0 and 2 (inclusively), where values close to 0 represent a higher
chance of been homozygous of the major allele, values close to 1 represent a
higher chance of been heterozygous, and values close to 2 represent a higher
chance of been homozygous of the minor allele.</p>
<p>The following example shows two lines of the <code class="docutils literal notranslate"><span class="pre">.dosage</span></code> file.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>22   16058758        rs7289830       C       A       0.0     0.0     1.0     ...
22   16087621        rs6423472       A       G       1.0     2.0     1.0     ...
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Dosage values computed from probabilities that are below the quality
threshold (specified by the <code class="docutils literal notranslate"><span class="pre">--prob</span></code> option) will have a missing value of
<code class="docutils literal notranslate"><span class="pre">nan</span></code>.</p>
</div>
</div>
<div class="section" id="calls-file">
<h3><code class="docutils literal notranslate"><span class="pre">.calls</span></code> file<a class="headerlink" href="#calls-file" title="Permalink to this headline">¶</a></h3>
<p>This file contains the hard calls computed from the <em>impute2</em> probabilities. It
has the same format as a transposed pedfile (from <em>Plink</em>). The general
structure of the file contains the following columns (which are tabulation
separated): the chromosome, the marker name, the genetic position, the genomic
location, and the hard calls.</p>
<p>The following example shows two lines of the <code class="docutils literal notranslate"><span class="pre">.calls</span></code> file.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>22   rs7289830       0       16058758        A A     A A     C A     ...
22   rs6423472       0       16087621        A G     A A     A G     ...
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Hard calls computed from probabilities that are below the quality threshold
(specified by the <code class="docutils literal notranslate"><span class="pre">--prob</span></code> option) will have a missing value of <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">0</span></code>.</p>
</div>
</div>
<div class="section" id="binary-plink-files">
<h3>Binary <em>Plink</em> files<a class="headerlink" href="#binary-plink-files" title="Permalink to this headline">¶</a></h3>
<p>A set of three files are created (<em>i.e.</em> <code class="docutils literal notranslate"><span class="pre">.bed</span></code>, <code class="docutils literal notranslate"><span class="pre">.bim</span></code> and <code class="docutils literal notranslate"><span class="pre">.fam</span></code> files.
These represents binary <em>Plink</em> files containing hard calls.</p>
</div>
</div>
<div class="section" id="usage">
<span id="extract-tut-usage"></span><h2><a class="toc-backref" href="#id5">Usage</a><a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>The following command will display the documentation for the extraction
analysis in the console:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>impute2-extractor --help
<span class="go">usage: impute2-extractor [-h] [-v] [--debug] --impute2 FILE [--index]</span>
<span class="go">                         [--out PREFIX] [--format FORMAT [FORMAT ...]]</span>
<span class="go">                         [--long] [--prob FLOAT] [--extract FILE]</span>
<span class="go">                         [--genomic CHR:START-END] [--maf FLOAT]</span>
<span class="go">                         [--rate FLOAT] [--info FLOAT]</span>

<span class="go">Extract imputed markers located in a specific genomic region. This script is</span>
<span class="go">part of the &#39;genipe&#39; package, version 1.4.2.</span>

<span class="go">optional arguments:</span>
<span class="go">  -h, --help            show this help message and exit</span>
<span class="go">  -v, --version         show program&#39;s version number and exit</span>
<span class="go">  --debug               set the logging level to debug</span>

<span class="go">Input Files:</span>
<span class="go">  --impute2 FILE        The output from IMPUTE2.</span>

<span class="go">Indexation Options:</span>
<span class="go">  --index               Only perform the indexation.</span>

<span class="go">Output Options:</span>
<span class="go">  --out PREFIX          The prefix of the output files. [impute2_extractor]</span>
<span class="go">  --format FORMAT [FORMAT ...]</span>
<span class="go">                        The output format. Can specify either &#39;impute2&#39; for</span>
<span class="go">                        probabilities (same as impute2 format, i.e. 3 values</span>
<span class="go">                        per sample), &#39;dosage&#39; for dosage values (one value</span>
<span class="go">                        between 0 and 2 by sample), &#39;calls&#39; for hard calls, or</span>
<span class="go">                        &#39;bed&#39; for Plink binary format (with hard calls).</span>
<span class="go">                        [&#39;impute2&#39;]</span>
<span class="go">  --long                Write the output file in the long format (one line per</span>
<span class="go">                        sample per marker). This option is only compatible</span>
<span class="go">                        with the &#39;calls&#39; and &#39;dosage&#39; format (option &#39;--</span>
<span class="go">                        format&#39;).</span>
<span class="go">  --prob FLOAT          The probability threshold used when creating a file in</span>
<span class="go">                        the dosage or call format. [0.9]</span>

<span class="go">Extraction Options:</span>
<span class="go">  --extract FILE        File containing marker names to extract.</span>
<span class="go">  --genomic CHR:START-END</span>
<span class="go">                        The range to extract (e.g. 22 1000000 1500000). Can be</span>
<span class="go">                        use in combination with &#39;--rate&#39;, &#39;--maf&#39; and &#39;--</span>
<span class="go">                        info&#39;.</span>
<span class="go">  --maf FLOAT           Extract markers with a minor allele frequency equal or</span>
<span class="go">                        higher than the specified threshold. Can be use in</span>
<span class="go">                        combination with &#39;--rate&#39;, &#39;--info&#39; and &#39;--genomic&#39;.</span>
<span class="go">  --rate FLOAT          Extract markers with a completion rate equal or higher</span>
<span class="go">                        to the specified threshold. Can be use in combination</span>
<span class="go">                        with &#39;--maf&#39;, &#39;--info&#39; and &#39;--genomic&#39;.</span>
<span class="go">  --info FLOAT          Extract markers with an information equal or higher to</span>
<span class="go">                        the specified threshold. Can be use in combination</span>
<span class="go">                        with &#39;--maf&#39;, &#39;--rate&#39; and &#39;--genomic&#39;.</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When using the <code class="docutils literal notranslate"><span class="pre">--index</span></code> option, only the indexation (of files without an
index) will be performed.</p>
</div>
</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Site extraction</a><ul>
<li><a class="reference internal" href="#input-files">Input files</a></li>
<li><a class="reference internal" href="#executing-the-extraction">Executing the extraction</a><ul>
<li><a class="reference internal" href="#extraction-by-id">Extraction by ID</a></li>
<li><a class="reference internal" href="#extraction-by-characteristics">Extraction by characteristics</a></li>
</ul>
</li>
<li><a class="reference internal" href="#output-files">Output files</a><ul>
<li><a class="reference internal" href="#impute2-file"><code class="docutils literal notranslate"><span class="pre">.impute2</span></code> file</a></li>
<li><a class="reference internal" href="#dosage-file"><code class="docutils literal notranslate"><span class="pre">.dosage</span></code> file</a></li>
<li><a class="reference internal" href="#calls-file"><code class="docutils literal notranslate"><span class="pre">.calls</span></code> file</a></li>
<li><a class="reference internal" href="#binary-plink-files">Binary <em>Plink</em> files</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">Usage</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_logistic.html"
                        title="previous chapter">Logistic regression</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="../execution_time.html"
                        title="next chapter">Statistical Analysis Execution Time</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/tutorials/tutorial_extract.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../execution_time.html" title="Statistical Analysis Execution Time"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_logistic.html" title="Logistic regression"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">genipe 1.5.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorials.html" >Short Tutorials</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Site extraction</a></li> 
      </ul>
    </div>


    <div class="footer" role="contentinfo">
        &#169; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 3.4.3.
    </div>

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-81009922-1', 'auto');
  ga('send', 'pageview');

</script>


  </body>
</html>