

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>SKAT analysis &mdash; genipe 1.3.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.3.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/bizstyle.js"></script>
    <link rel="top" title="genipe 1.3.0 documentation" href="../index.html" />
    <link rel="up" title="Short Tutorials" href="../tutorials.html" />
    <link rel="next" title="Linear regression" href="tutorial_linear.html" />
    <link rel="prev" title="Survival analysis" href="tutorial_cox.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_linear.html" title="Linear regression"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_cox.html" title="Survival analysis"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">genipe 1.3.0 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorials.html" accesskey="U">Short Tutorials</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">SKAT analysis</a><ul>
<li><a class="reference internal" href="#input-files-generated-by-genipe">Input files generated by genipe</a><ul>
<li><a class="reference internal" href="#impute2">Impute2</a></li>
<li><a class="reference internal" href="#samples-file">Samples file</a></li>
<li><a class="reference internal" href="#good-sites-file">Good sites file</a></li>
</ul>
</li>
<li><a class="reference internal" href="#creating-a-snp-sets-file">Creating a SNP sets file</a></li>
<li><a class="reference internal" href="#format-for-the-phenotypes-file">Format for the phenotypes file</a></li>
<li><a class="reference internal" href="#running-the-script">Running the script</a></li>
<li><a class="reference internal" href="#results">Results</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_cox.html"
                        title="previous chapter">Survival analysis</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_linear.html"
                        title="next chapter">Linear regression</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/tutorials/tutorial_SKAT.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="contents topic" id="quick-navigation">
<p class="topic-title first">Quick navigation</p>
<ul class="simple">
<li><a class="reference internal" href="#skat-analysis" id="id2">SKAT analysis</a><ul>
<li><a class="reference internal" href="#input-files-generated-by-genipe" id="id3">Input files generated by genipe</a></li>
<li><a class="reference internal" href="#creating-a-snp-sets-file" id="id4">Creating a SNP sets file</a></li>
<li><a class="reference internal" href="#format-for-the-phenotypes-file" id="id5">Format for the phenotypes file</a></li>
<li><a class="reference internal" href="#running-the-script" id="id6">Running the script</a></li>
<li><a class="reference internal" href="#results" id="id7">Results</a></li>
<li><a class="reference internal" href="#usage" id="id8">Usage</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="skat-analysis">
<h1><a class="toc-backref" href="#id2">SKAT analysis</a><a class="headerlink" href="#skat-analysis" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="http://www.hsph.harvard.edu/skat/">SKAT</a> is a very popular test for the
association of sets of rare and common variants. In <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a>, we
provide straightforward bindings over SKAT to make it easy to perform parallel
SNP-set analysis of imputed variants.</p>
<p>This tutorial will walk through a full example of an analysis while describing
the expected input file formats and the resulting output files at every
step.</p>
<div class="section" id="input-files-generated-by-genipe">
<span id="skat-tut-1"></span><h2><a class="toc-backref" href="#id3">Input files generated by genipe</a><a class="headerlink" href="#input-files-generated-by-genipe" title="Permalink to this headline">¶</a></h2>
<div class="section" id="impute2">
<h3>Impute2<a class="headerlink" href="#impute2" title="Permalink to this headline">¶</a></h3>
<p>After running the pipeline, you should have an <cite>Impute2</cite> file containing
genotype probabilities for all the imputed variants. The general structure of
the <cite>Impute2</cite> file contains information on the variant: the chromosome, name,
position, first and second alleles. The subsequent columns correspond to the
probabilities for the AA, AB and BB genotypes for all the variants. Hence, a
single variant is represented by three columns in <cite>Impute2</cite> files.</p>
<p>The following example shows a single line of the <cite>Impute2</cite> file where column
identifiers were added to help explain the file structure.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">A</span> <span class="n">B</span>            <span class="n">C</span>        <span class="n">D</span> <span class="n">E</span> <span class="mi">1</span> <span class="mi">2</span> <span class="mi">3</span> <span class="mi">1</span> <span class="mi">2</span> <span class="mi">3</span> <span class="mi">1</span> <span class="mi">2</span> <span class="mi">3</span> <span class="mi">1</span>     <span class="mi">2</span>     <span class="mi">3</span>
<span class="mi">3</span> <span class="mi">3</span><span class="p">:</span><span class="mi">20109120</span><span class="p">:</span><span class="n">C</span> <span class="mi">20109120</span> <span class="n">A</span> <span class="n">C</span> <span class="mi">1</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">1</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">1</span> <span class="mf">0.998</span> <span class="mf">0.002</span> <span class="mi">0</span>
<span class="mi">3</span> <span class="n">rs79511213</span><span class="p">:</span><span class="n">T</span> <span class="mi">20202955</span> <span class="n">C</span> <span class="n">T</span> <span class="mi">1</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">1</span> <span class="mi">0</span> <span class="mi">1</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mf">0.989</span> <span class="mf">0.011</span> <span class="mi">0</span>
<span class="o">...</span>
</pre></div>
</div>
<p>The columns for the previous examples go as follows:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">A</span></code> The chromosome.</li>
<li><code class="docutils literal"><span class="pre">B</span></code> The variant name.</li>
<li><code class="docutils literal"><span class="pre">C</span></code> The genomic position of the variant.</li>
<li><code class="docutils literal"><span class="pre">D</span></code> The &#8220;A1&#8221; allele. This is not necessarily the reference or alternative
allele.</li>
<li><code class="docutils literal"><span class="pre">E</span></code> The &#8220;A2&#8221; allele. This is not necessarily the reference or alternative
allele.</li>
<li><code class="docutils literal"><span class="pre">1</span></code> All those columns represent the probability of being a A1-A1
homozygous. It is noticeable that the columns 1, 2 and 3 alternate. This is
because probabilities are given for all the variants in order. In other
words, the probabilities are read as triplets.</li>
<li><code class="docutils literal"><span class="pre">2</span></code> Columns representing the A1-A2 heterozygous.</li>
<li><code class="docutils literal"><span class="pre">3</span></code> Columns representing the A2-A2 heterozygous.</li>
</ul>
<p>Also note that identifying the minor and major alleles is done internally by
computing the minor allele frequency (MAF) and assigning the rarest allele to
be the minor allele.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Real <cite>Impute2</cite> files do not have the <cite>A, B, C, ...</cite> row. It was only added
to facilitate the description of the format.</p>
</div>
</div>
<div class="section" id="samples-file">
<h3>Samples file<a class="headerlink" href="#samples-file" title="Permalink to this headline">¶</a></h3>
<p>This file is generated by <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a> and has the <code class="docutils literal"><span class="pre">.sample</span></code> extension.
It greatly resembles the PLINK <code class="docutils literal"><span class="pre">fam</span></code> file for those who are familiar with
this format. Specifically, it contains the samples that are included in the
<code class="docutils literal"><span class="pre">Impute2</span></code> file <strong>in the correct order</strong>. It is needed to correctly interpret
the sample described by the probability triplets described earlier.</p>
<p>Specifically, the format is as follow:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">ID_1</span> <span class="n">ID_2</span> <span class="n">missing</span> <span class="n">father</span> <span class="n">mother</span> <span class="n">sex</span> <span class="n">plink_pheno</span>
<span class="mi">0</span> <span class="mi">0</span> <span class="mi">0</span> <span class="n">D</span> <span class="n">D</span> <span class="n">D</span> <span class="n">B</span>
<span class="n">SAMPLE1</span> <span class="n">SAMPLE1</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">1</span> <span class="o">-</span><span class="mi">9</span>
<span class="n">SAMPLE2</span> <span class="n">SAMPLE2</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">2</span> <span class="o">-</span><span class="mi">9</span>
<span class="n">SAMPLE3</span> <span class="n">SAMPLE3</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">2</span> <span class="o">-</span><span class="mi">9</span>
<span class="n">SAMPLE4</span> <span class="n">SAMPLE4</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">1</span> <span class="o">-</span><span class="mi">9</span>
<span class="o">...</span>
</pre></div>
</div>
<p>This time, the <strong>first rows are part of the format</strong>. The name of the first two
columns is important for the analysis to work, and omitting the second row will
result in missing samples in the analysis.</p>
<p>The attentive reader will notice that this <cite>sample</cite> file, when compared to the
previously described <cite>Impute2</cite> file yields the following dosage matrix (if we
assume that the A1 is the major allele, which is not true in practice):</p>
<table border="1" class="docutils" id="id1">
<caption><span class="caption-text">Dosage and hard-calls derived from the <cite>Impute2</cite> and <cite>.sample</cite> files
       described in this tutorial.</span><a class="headerlink" href="#id1" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="27%" />
<col width="34%" />
<col width="39%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&nbsp;</th>
<th class="head">3:20109120:C</th>
<th class="head">rs79511213:T</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>SAMPLE1</td>
<td>0 (A1-A1)</td>
<td>0 (A1-A1)</td>
</tr>
<tr class="row-odd"><td>SAMPLE2</td>
<td>0 (A1-A1)</td>
<td>1 (A1-A2)</td>
</tr>
<tr class="row-even"><td>SAMPLE3</td>
<td>2 (A2-A2)</td>
<td>0 (A1-A1)</td>
</tr>
<tr class="row-odd"><td>SAMPLE4</td>
<td>0.002 (A1-A1)</td>
<td>0.011 (A1-A1)</td>
</tr>
</tbody>
</table>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Once again, note that for simplicity this example assumes that the A1 is
the major allele, which is often not true in practice.</p>
</div>
</div>
<div class="section" id="good-sites-file">
<h3>Good sites file<a class="headerlink" href="#good-sites-file" title="Permalink to this headline">¶</a></h3>
<p>The <cite>.good_sites</cite> file is also generated by <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a>. It contains a
single column representing the name of the variants that passed the quality
thresholds.</p>
<p>Only the variants in this file will be considered by the <code class="docutils literal"><span class="pre">imputed-stats</span></code>
tool.</p>
</div>
</div>
<div class="section" id="creating-a-snp-sets-file">
<span id="skat-tut-2"></span><h2><a class="toc-backref" href="#id4">Creating a SNP sets file</a><a class="headerlink" href="#creating-a-snp-sets-file" title="Permalink to this headline">¶</a></h2>
<p>SKAT is based on the analysis of a &#8220;SNP set&#8221;. This is simply an arbitrary set
of variants that is created by the user. All of the variants in a same SNP set
will be analyzed together by SKAT.</p>
<p>It is also noteworthy that adding neutral variants to a SNP set is generally
avoided as much as possible, as it is known to decrease SKAT&#8217;s power, even if
true deleterious variants are present in the SNP set. Describing the current
best practices to create SNP sets is out of the scope of this tutorial, but
possible approaches consist of testing variants in a same gene together, using
variant prioritization tools to select interesting candidates or to group
variants according to a biological pathway of interest.</p>
<p>The expected format for SNP sets is very straightforward. A tab separated file
with a single header row containing the following fields: <code class="docutils literal"><span class="pre">variant</span></code>,
<code class="docutils literal"><span class="pre">snp_set</span></code> is required. An optional third field identified by the <code class="docutils literal"><span class="pre">weight</span></code>
header can also be provided. Concretely, the expected file looks like this:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">variant</span>         <span class="n">snp_set</span>
<span class="mi">3</span><span class="p">:</span><span class="mi">20109120</span><span class="p">:</span><span class="n">C</span>    <span class="n">SET1</span>
<span class="n">exm295293</span>       <span class="n">SET1</span>
<span class="mi">3</span><span class="p">:</span><span class="mi">20225460</span><span class="p">:</span><span class="n">T</span>    <span class="n">SET1</span>
<span class="n">rs143768051</span><span class="p">:</span><span class="n">T</span>   <span class="n">SET1</span>
<span class="n">rs79511213</span><span class="p">:</span><span class="n">T</span>    <span class="n">SET2</span>
<span class="n">rs183736430</span><span class="p">:</span><span class="n">T</span>   <span class="n">SET2</span>
<span class="n">rs375511922</span><span class="p">:</span><span class="n">G</span>   <span class="n">SET2</span>
<span class="n">rs369840848</span><span class="p">:</span><span class="n">G</span>   <span class="n">SET2</span>
</pre></div>
</div>
<p>Here, two gene sets are defined named <cite>SET1</cite> and <cite>SET2</cite>. They both contain 4
variants, which is a small number for a real life application, especially if
the variants are rare.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Don&#8217;t forget that this file has to be tab (t) delimited!</p>
</div>
<p>Another example using weights is presented here:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">variant</span>         <span class="n">snp_set</span>  <span class="n">weight</span>
<span class="mi">3</span><span class="p">:</span><span class="mi">20109120</span><span class="p">:</span><span class="n">C</span>    <span class="n">SET1</span>     <span class="mf">1.4075226244343892</span>
<span class="n">exm295293</span>       <span class="n">SET1</span>     <span class="mf">1.8227675243926944</span>
<span class="mi">3</span><span class="p">:</span><span class="mi">20225460</span><span class="p">:</span><span class="n">T</span>    <span class="n">SET1</span>     <span class="mf">1.7138009049773757</span>
<span class="n">rs143768051</span><span class="p">:</span><span class="n">T</span>   <span class="n">SET1</span>     <span class="mf">1.5197343711318112</span>
<span class="n">rs79511213</span><span class="p">:</span><span class="n">T</span>    <span class="n">SET2</span>     <span class="mf">0.6643099547511312</span>
<span class="n">rs183736430</span><span class="p">:</span><span class="n">T</span>   <span class="n">SET2</span>     <span class="mf">0.1363741358618047</span>
<span class="n">rs375511922</span><span class="p">:</span><span class="n">G</span>   <span class="n">SET2</span>     <span class="mf">0.4431278280542987</span>
<span class="n">rs369840848</span><span class="p">:</span><span class="n">G</span>   <span class="n">SET2</span>     <span class="mf">0.1171662895927601</span>
<span class="n">rs202028804</span><span class="p">:</span><span class="n">C</span>   <span class="n">SET2</span>     <span class="mf">0.5328902714932127</span>
</pre></div>
</div>
<p>The only difference is the added <cite>weight</cite> field that is used by SKAT. By
default, SKAT uses a weighting scheme that assigns larger prior weights to rare
variants. If you want to use custom weights based on something la
deleteriousness prediction, you need to generate the weights manually and
provide them in the SNP set file as shown here.</p>
</div>
<div class="section" id="format-for-the-phenotypes-file">
<span id="skat-tut-3"></span><h2><a class="toc-backref" href="#id5">Format for the phenotypes file</a><a class="headerlink" href="#format-for-the-phenotypes-file" title="Permalink to this headline">¶</a></h2>
<p>The last required file is the phenotype file. The latter contains information
on the phenotype of interest for all the samples included in the analysis. The
expected format is a tab delimited file containing a user-specified identifier
column (through the <code class="docutils literal"><span class="pre">--sample-column</span></code> argument) and a phenotype column (
represented by the <code class="docutils literal"><span class="pre">--pheno-name</span></code> argument).</p>
<p>The order of the samples is not important, as long as the identifier are
consistent with the <code class="docutils literal"><span class="pre">.sample</span></code> file.</p>
<p>The following file is a valid example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Sample</span>  <span class="n">Hypertension</span> <span class="n">SystolicBP</span> <span class="n">Age</span>
<span class="n">SAMPLE1</span> <span class="mi">1</span>            <span class="mf">163.21</span>     <span class="mi">63</span>
<span class="n">SAMPLE2</span> <span class="mi">0</span>            <span class="mf">122.13</span>     <span class="mi">21</span>
<span class="n">SAMPLE3</span> <span class="mi">0</span>            <span class="mf">119.01</span>     <span class="mi">42</span>
<span class="n">SAMPLE4</span> <span class="mi">1</span>            <span class="mf">158.84</span>     <span class="mi">79</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This file has to be tab (t) delimited!</p>
</div>
<p>As you can see, both continuous an discrete traits can be included in this
file, as well as covariates. The covariates are included in the analysis using
the <code class="docutils literal"><span class="pre">--covar</span></code> argument and the variable type of the outcome is given using
the <code class="docutils literal"><span class="pre">--outcome-type</span></code> argument which can be set to <cite>discrete</cite> or <cite>continuous</cite>.</p>
</div>
<div class="section" id="running-the-script">
<span id="skat-tut-4"></span><h2><a class="toc-backref" href="#id6">Running the script</a><a class="headerlink" href="#running-the-script" title="Permalink to this headline">¶</a></h2>
<p>Once all of this is ready, you can finally run the script. A sample command
for the analysis described throughout this tutorial is the following:</p>
<div class="highlight-console"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="go">imputed-stats skat \</span>
<span class="go">    --impute2 data/genipe_is_great.impute2 \</span>
<span class="go">    --sample data/genipe_is_great.sample \</span>
<span class="go">    --pheno data/phenotypes.txt \</span>
<span class="go">    --sample-column Sample \</span>
<span class="go">    --pheno-name SystolicBP \</span>
<span class="go">    --covar Age \</span>
<span class="go">    --extract-sites data/genipe_is_great.good_sites \</span>
<span class="go">    --nb-process 4 \</span>
<span class="go">    --out my_skat_analysis \</span>
<span class="go">    --snp-sets carefully_crafted_snp_sets.txt \</span>
<span class="go">    --outcome-type continuous \</span>
</pre></div>
</td></tr></table></div>
<p>The line by line explanation of this command is as follows:</p>
<ol class="arabic simple">
<li>We are calling the program with the <code class="docutils literal"><span class="pre">skat</span></code> option to specify the type of
analysis. Alternative analysis types are presented in other tutorials, but
they include linear and logistic regression as well as Cox survival
analysis.</li>
<li>The <cite>Impute2</cite> file as previously described. This contains the probability
for every allele of every variant. It is generated by <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a>.</li>
<li>The <cite>sample</cite> contained the sample identifiers with an ordering consistent
with the <cite>Impute2</cite> file. It was generated by <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a>.</li>
<li>The <cite>phenotype</cite> is a tab delimited file containing phenotypic information
for all the samples. Covariates are also included in this file.</li>
<li>The sample column is the identifier for the column containing sample IDs in
the phenotype file.</li>
<li>The phenotype column is the identifier for the outcome for this analysis. In
our case, we are using the quantitative trait <cite>SystolicBP</cite> representing the
systolic blood pressure in this example.</li>
<li>The covariates are provided as a comma separated list of column names from
the phenotype file. In our case, we&#8217;re only using the <cite>Age</cite> column, but
adding covariates is easy. As an example <cite>&#8211;covar Age,Hypertension</cite> would
add the <cite>Hypertension</cite> column as a covariate.</li>
<li>A list of variants to keep for the analysis is provided here. A good choice
would be to simply use the <cite>.good_sites</cite> file generated by py:mod:<cite>genipe</cite>,
but additional QC steps could be used to further narrow this list.</li>
<li>The <code class="docutils literal"><span class="pre">--nb-process</span></code> argument determines how many simultaneous SKAT analyses
will be started in parallel. This value should be a multiple of the number
of the number of SNP sets or of the number of CPUs on your machine as
parallelism only occurs at the gene set level.</li>
<li>The output prefix for the analysis results.</li>
<li>The SNP set file that was created by the user.</li>
<li>The outcome type. This should be set to either &#8216;continuous&#8217; or &#8216;discrete&#8217;
depending on the type of dependent variable under study.</li>
</ol>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Also note that the SKAT-O algorithm can be used by using the <code class="docutils literal"><span class="pre">--skat-o</span></code>
flag.</p>
</div>
</div>
<div class="section" id="results">
<span id="skat-tut-5"></span><h2><a class="toc-backref" href="#id7">Results</a><a class="headerlink" href="#results" title="Permalink to this headline">¶</a></h2>
<p>After running this analysis, a file named <cite>my_skat_analysis.skat.dosage</cite> should
appear. This file will contain an association p-value for every one of the
specified SNP sets.</p>
<p>You can also verify the <cite>my_skat_analysis.log</cite> file to see if <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a>
or <cite>SKAT</cite> generated any warnings.</p>
<p>Finally, a directory containing <cite>.csv</cite> files and an <cite>R</cite> script should also be
visible. This directory contains everything that is needed to run the analysis
in R. You can verify that the generated R script is consistent with your
expectations and that the analysis is correct. If the file sizes are
manageable, you can also archive them to insure reproducibility of the
analysis.</p>
<p>If you are unsatisfied about the generated R script, you can download the
<a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a> source and modify the
<a class="reference external" href="https://github.com/pgxcentre/genipe/blob/master/genipe/script_templates/run_skat.R">run_skat.R</a>
file. This is a template rendered and launched by <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a> to conduct
the analysis by using the regular R SKAT package. It is easy to extend this
file to customize your analysis or to write new ones.</p>
<p>If you have any questions or problems, don&#8217;t hesitate to contact us through our
<a class="reference external" href="https://github.com/pgxcentre">Github page</a> or by creating a new issue in the
<a class="reference external" href="https://github.com/pgxcentre/genipe/issues">Issue tracker</a>.</p>
<p>We are also very open to suggestions and improvements. If you have developed a
new interesting feature, please send us a push request from Github.</p>
</div>
<div class="section" id="usage">
<span id="skat-tut-6"></span><h2><a class="toc-backref" href="#id8">Usage</a><a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>The following command will display the documentation for the SKAT analysis in
the console:</p>
<div class="highlight-console"><div class="highlight"><pre><span></span><span class="gp">$</span> imputed-stats skat --help
<span class="go">usage: imputed-stats skat [-h] [-v] [--debug] --impute2 FILE --sample FILE</span>
<span class="go">                          --pheno FILE [--extract-sites FILE] [--out FILE]</span>
<span class="go">                          [--nb-process INT] [--nb-lines INT] [--chrx]</span>
<span class="go">                          [--gender-column NAME] [--scale INT] [--prob FLOAT]</span>
<span class="go">                          [--maf FLOAT] [--covar NAME] [--categorical NAME]</span>
<span class="go">                          [--missing-value NAME] [--sample-column NAME]</span>
<span class="go">                          [--interaction NAME] --snp-sets FILE --outcome-type</span>
<span class="go">                          {continuous,discrete} [--skat-o] --pheno-name NAME</span>

<span class="go">Uses the SKAT R package to analyze user defined gene sets. This script is part</span>
<span class="go">of the &#39;genipe&#39; package, version 1.3.0.</span>

<span class="go">optional arguments:</span>
<span class="go">  -h, --help            show this help message and exit</span>
<span class="go">  -v, --version         show program&#39;s version number and exit</span>
<span class="go">  --debug               set the logging level to debug</span>

<span class="go">Input Files:</span>
<span class="go">  --impute2 FILE        The output from IMPUTE2.</span>
<span class="go">  --sample FILE         The sample file (the order should be the same as in</span>
<span class="go">                        the IMPUTE2 files).</span>
<span class="go">  --pheno FILE          The file containing phenotypes and co variables.</span>
<span class="go">  --extract-sites FILE  A list of sites to extract for analysis (optional).</span>

<span class="go">Output Options:</span>
<span class="go">  --out FILE            The prefix for the output files. [imputed_stats]</span>

<span class="go">General Options:</span>
<span class="go">  --nb-process INT      The number of process to use. [1]</span>
<span class="go">  --nb-lines INT        The number of line to read at a time. [1000]</span>
<span class="go">  --chrx                The analysis is performed for the non pseudo-autosomal</span>
<span class="go">                        region of the chromosome X (male dosage will be</span>
<span class="go">                        divided by 2 to get values [0, 0.5] instead of [0, 1])</span>
<span class="go">                        (males are coded as 1 and option &#39;--gender-column&#39;</span>
<span class="go">                        should be used).</span>
<span class="go">  --gender-column NAME  The name of the gender column (use to exclude samples</span>
<span class="go">                        with unknown gender (i.e. not 1, male, or 2, female).</span>
<span class="go">                        If gender not available, use &#39;None&#39;. [Gender]</span>

<span class="go">Dosage Options:</span>
<span class="go">  --scale INT           Scale dosage so that values are in [0, n] (possible</span>
<span class="go">                        values are 1 (no scaling) or 2). [2]</span>
<span class="go">  --prob FLOAT          The minimal probability for which a genotype should be</span>
<span class="go">                        considered. [&gt;=0.9]</span>
<span class="go">  --maf FLOAT           Minor allele frequency threshold for which marker will</span>
<span class="go">                        be skipped. [&lt;0.01]</span>

<span class="go">Phenotype Options:</span>
<span class="go">  --covar NAME          The co variable names (in the phenotype file),</span>
<span class="go">                        separated by coma.</span>
<span class="go">  --categorical NAME    The name of the variables that are categorical (note</span>
<span class="go">                        that the gender is always categorical). The variables</span>
<span class="go">                        are separated by coma.</span>
<span class="go">  --missing-value NAME  The missing value in the phenotype file.</span>
<span class="go">  --sample-column NAME  The name of the sample ID column (in the phenotype</span>
<span class="go">                        file). [sample_id]</span>
<span class="go">  --interaction NAME    Add an interaction between the genotype and this</span>
<span class="go">                        variable.</span>

<span class="go">SKAT Options:</span>
<span class="go">  --snp-sets FILE       A file indicating a snp_set and an optional weight for</span>
<span class="go">                        every variant.</span>
<span class="go">  --outcome-type {continuous,discrete}</span>
<span class="go">                        The variable type for the outcome. This will be passed</span>
<span class="go">                        to SKAT.</span>
<span class="go">  --skat-o              By default, the regular SKAT is used. Setting this</span>
<span class="go">                        flag will use the SKAT-O algorithm instead.</span>
<span class="go">  --pheno-name NAME     The phenotype.</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_linear.html" title="Linear regression"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_cox.html" title="Survival analysis"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">genipe 1.3.0 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorials.html" >Short Tutorials</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.4.
    </div>
  </body>
</html>