

<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Genome-wide imputation pipeline &#8212; genipe 1.4.2 documentation</title>
    <link rel="stylesheet" href="../_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/bizstyle.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Linear mixed effects model" href="tutorial_mixedlm.html" />
    <link rel="prev" title="Short Tutorials" href="../tutorials.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_mixedlm.html" title="Linear mixed effects model"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../tutorials.html" title="Short Tutorials"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">genipe 1.4.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorials.html" accesskey="U">Short Tutorials</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Genome-wide imputation pipeline</a><ul>
<li><a class="reference internal" href="#preparing-the-data">Preparing the data</a></li>
<li><a class="reference internal" href="#input-file-summary">Input file summary</a></li>
<li><a class="reference internal" href="#executing-the-pipeline">Executing the pipeline</a></li>
<li><a class="reference internal" href="#compiling-the-report">Compiling the report</a></li>
<li><a class="reference internal" href="#output-files">Output files</a><ul>
<li><a class="reference internal" href="#genipe-directory"><code class="docutils literal notranslate"><span class="pre">genipe</span></code> directory</a></li>
<li><a class="reference internal" href="#genipe-chrn-directories"><code class="docutils literal notranslate"><span class="pre">genipe/chrN</span></code> directories</a></li>
<li><a class="reference internal" href="#genipe-chrn-final-impute2-directories"><code class="docutils literal notranslate"><span class="pre">genipe/chrN/final_impute2</span></code> directories</a></li>
<li><a class="reference internal" href="#genipe-missing-directory"><code class="docutils literal notranslate"><span class="pre">genipe/missing</span></code> directory</a></li>
<li><a class="reference internal" href="#genipe-report-directory"><code class="docutils literal notranslate"><span class="pre">genipe/report</span></code> directory</a></li>
</ul>
</li>
<li><a class="reference internal" href="#using-genipe-on-a-server">Using genipe on a server</a><ul>
<li><a class="reference internal" href="#drmaa-configuration">DRMAA configuration</a></li>
<li><a class="reference internal" href="#drmaa-preamble">DRMAA preamble</a></li>
</ul>
</li>
<li><a class="reference internal" href="#data-in-more-details">Data in more details</a><ul>
<li><a class="reference internal" href="#required-softwares">Required softwares</a></li>
<li><a class="reference internal" href="#input-files">Input files</a><ul>
<li><a class="reference internal" href="#genotypes">Genotypes</a></li>
<li><a class="reference internal" href="#reference-panels">Reference panels</a></li>
<li><a class="reference internal" href="#human-reference-optional">Human reference (optional)</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../tutorials.html"
                        title="previous chapter">Short Tutorials</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_mixedlm.html"
                        title="next chapter">Linear mixed effects model</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/tutorials/tutorial_genipe.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="contents topic" id="quick-navigation">
<p class="topic-title first">Quick navigation</p>
<ul class="simple">
<li><a class="reference internal" href="#genome-wide-imputation-pipeline" id="id2">Genome-wide imputation pipeline</a><ul>
<li><a class="reference internal" href="#preparing-the-data" id="id3">Preparing the data</a></li>
<li><a class="reference internal" href="#input-file-summary" id="id4">Input file summary</a></li>
<li><a class="reference internal" href="#executing-the-pipeline" id="id5">Executing the pipeline</a></li>
<li><a class="reference internal" href="#compiling-the-report" id="id6">Compiling the report</a></li>
<li><a class="reference internal" href="#output-files" id="id7">Output files</a></li>
<li><a class="reference internal" href="#using-genipe-on-a-server" id="id8">Using genipe on a server</a></li>
<li><a class="reference internal" href="#data-in-more-details" id="id9">Data in more details</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="genome-wide-imputation-pipeline">
<span id="genipe-tut-page"></span><h1><a class="toc-backref" href="#id2">Genome-wide imputation pipeline</a><a class="headerlink" href="#genome-wide-imputation-pipeline" title="Permalink to this headline">¶</a></h1>
<p>High-throughput genotyping platforms can assess up to five million variations
in thousands of samples. Depending on the genotyped data and the reference
panel used, genome-wide imputation tools can infer genotypes for more than 38
million variants (single nucleotides, insertions and deletions). Unfortunately,
genome-wide imputation requires high computation power and multiple data
processing steps.</p>
<p>The <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genipe</span></code></a> pipeline automates the different steps for pre-phasing and
imputation for genome-wide data. The pipeline follows the guideline described
by IMPUTE2’s best practices when analyzing genome-wide data (described
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html#prephasing">by IMPUTE2</a>
and <a class="reference external" href="http://www.shapeit.fr/pages/m03_phasing/imputation.html">by SHAPEIT</a>).</p>
<div class="section" id="preparing-the-data">
<h2><a class="toc-backref" href="#id3">Preparing the data</a><a class="headerlink" href="#preparing-the-data" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genipe</span></code></a> module provides a script to prepare all the required files
for this tutorial. It will download the appropriate binary files for Plink,
IMPUTE2 and SHAPETIT, the genotypes of the study cohort to impute, the
IMPUTE2’s reference panels, and the indexed human reference (in <em>fasta</em>
format).</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">You should review
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html#download">IMPUTE2 licence</a> and
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#download">SHAPEIT licence</a>
to make sure that you have the right to use these tools.</p>
</div>
<p>To following command will execute the script (once the virtual Python
environment is activated, see <a class="reference internal" href="../installation.html#genipe-pyvenv-activation"><span class="std std-ref">Virtual environment activation</span></a> for more
details).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>genipe-tutorial
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The script will create a directory called <code class="docutils literal notranslate"><span class="pre">genipe_tutorial</span></code> in the home
directory. If another directory is required, the default path can be modify
using the <code class="docutils literal notranslate"><span class="pre">--tutorial-path</span></code> option.</p>
</div>
<p>To have more information about the input data, have a look at the
<a class="reference internal" href="#genipe-tut-more-details"><span class="std std-ref">Data in more details</span></a> section below. If the imputation is to be run on
a server using a job scheduler, have a look at the <a class="reference internal" href="#genipe-tut-drmaa"><span class="std std-ref">Using genipe on a server</span></a>
section.</p>
</div>
<div class="section" id="input-file-summary">
<h2><a class="toc-backref" href="#id4">Input file summary</a><a class="headerlink" href="#input-file-summary" title="Permalink to this headline">¶</a></h2>
<p>You should have the following directory structure:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$HOME/genipe_tutorial/
│
├── 1000GP_Phase3/
│   ├── 1000GP_Phase3_chr1.hap.gz
│   ├── 1000GP_Phase3_chr2.hap.gz
│   ├── ...
│   ├── 1000GP_Phase3_chr1.legend.gz
│   ├── 1000GP_Phase3_chr2.legend.gz
│   ├── ...
│   ├── 1000GP_Phase3.sample
│   ├── genetic_map_chr1_combined_b37.txt
│   ├── genetic_map_chr2_combined_b37.txt
│   └── ...
│
├── bin/
│   ├── impute2
│   ├── plink
│   └── shapeit
│
├── data/
│   ├── hapmap_CEU_r23a_hg19.bed
│   ├── hapmap_CEU_r23a_hg19.bim
│   └── hapmap_CEU_r23a_hg19.fam
│
├── genipe_config.ini  # OPTIONAL (--use-drmaa, --drmaa-config)
│
├── hg19/
│   ├── hg19.fasta
│   └── hg19.fasta.fai
│
└── preamble.txt     # OPTIONAL (--use-drmaa, --preamble)
</pre></div>
</div>
</div>
<div class="section" id="executing-the-pipeline">
<span id="genipe-tut-execute"></span><h2><a class="toc-backref" href="#id5">Executing the pipeline</a><a class="headerlink" href="#executing-the-pipeline" title="Permalink to this headline">¶</a></h2>
<p>Once all the input files are ready for analysis, you can finally execute the
pipeline. Make sure that the virtual Python environment was properly activated
(see <a class="reference internal" href="../installation.html#genipe-pyvenv-activation"><span class="std std-ref">Virtual environment activation</span></a> for more details).</p>
<p>When automatically preparing the data using the <code class="docutils literal notranslate"><span class="pre">genipe-tutorial</span></code> script, a
new script named <code class="docutils literal notranslate"><span class="pre">execute.sh</span></code> will be created to execute the analysis. Here
is the content of the <code class="docutils literal notranslate"><span class="pre">execute.sh</span></code> script.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/env bash</span>
<span class="c1"># Changing directory</span>
<span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial

<span class="c1"># Launching the imputation with genipe</span>
genipe-launcher <span class="se">\</span>
    --chrom autosomes <span class="se">\</span>
    --bfile <span class="nv">$HOME</span>/genipe_tutorial/data/hapmap_CEU_r23a_hg19 <span class="se">\</span>
    --shapeit-bin <span class="nv">$HOME</span>/genipe_tutorial/bin/shapeit <span class="se">\</span>
    --impute2-bin <span class="nv">$HOME</span>/genipe_tutorial/bin/impute2 <span class="se">\</span>
    --plink-bin <span class="nv">$HOME</span>/genipe_tutorial/bin/plink <span class="se">\</span>
    --reference <span class="nv">$HOME</span>/genipe_tutorial/hg19/hg19.fasta <span class="se">\</span>
    --hap-template <span class="nv">$HOME</span>/genipe_tutorial/1000GP_Phase3/1000GP_Phase3_chr<span class="o">{</span>chrom<span class="o">}</span>.hap.gz <span class="se">\</span>
    --legend-template <span class="nv">$HOME</span>/genipe_tutorial/1000GP_Phase3/1000GP_Phase3_chr<span class="o">{</span>chrom<span class="o">}</span>.legend.gz <span class="se">\</span>
    --map-template <span class="nv">$HOME</span>/genipe_tutorial/1000GP_Phase3/genetic_map_chr<span class="o">{</span>chrom<span class="o">}</span>_combined_b37.txt <span class="se">\</span>
    --sample-file <span class="nv">$HOME</span>/genipe_tutorial/1000GP_Phase3/1000GP_Phase3.sample <span class="se">\</span>
    --filtering-rules <span class="s1">&#39;ALL&lt;0.01&#39;</span> <span class="s1">&#39;ALL&gt;0.99&#39;</span> <span class="se">\</span>
    --thread <span class="m">4</span> <span class="se">\</span>
    --report-title <span class="s2">&quot;Tutorial&quot;</span> <span class="se">\</span>
    --report-number <span class="s2">&quot;Test Report&quot;</span>
</pre></div>
</div>
<p>When in the correct working directory, the following command should execute the
genome-wide imputation of the <em>HapMap</em> CEU dataset.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial

./execute.sh
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In the script <code class="docutils literal notranslate"><span class="pre">execute.sh</span></code>, the <code class="docutils literal notranslate"><span class="pre">--reference</span></code> option is optional and
might be removed.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">It is possible to compress IMPUTE2’s output files by adding the <code class="docutils literal notranslate"><span class="pre">--bgzip</span></code>
option (the <code class="docutils literal notranslate"><span class="pre">bgzip</span></code> software should be in the path).</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">It is possible to remove the <code class="docutils literal notranslate"><span class="pre">--shapeit-bin</span></code>, the <code class="docutils literal notranslate"><span class="pre">--impute2-bin</span></code> and/or
the <code class="docutils literal notranslate"><span class="pre">--plink-bin</span></code> options if the SHAPEIT, IMPUTE2 and/or the Plink
binaries are in  the <code class="docutils literal notranslate"><span class="pre">PATH</span></code> variable.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <code class="docutils literal notranslate"><span class="pre">--chrom</span> <span class="pre">autosomes</span></code> option will run the imputation only on the
autosomes. If all chromosomes are required (including chromosomes 23 and 25
(pseudo-autosomal region), then the <code class="docutils literal notranslate"><span class="pre">--chrom</span></code> option can be removed, and
the imputation reference files for chromosome 23 PAR and nonPAR are required
(see <a class="reference internal" href="../parameters/main_pipeline.html#params-chrom-x"><span class="std std-ref">this section</span></a> for more information).</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">It is possible to add extra options to SHAPEIT and IMPUTE2 using the
<code class="docutils literal notranslate"><span class="pre">--shapeit-extra</span></code> and <code class="docutils literal notranslate"><span class="pre">--impute2-extra</span></code> options, respectively.</p>
</div>
<p>The following table describes the options <strong>used by</strong> <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genipe</span></code></a> <strong>in the
previous command</strong> (see the <a class="reference internal" href="../index.html#genipe-usage"><span class="std std-ref">Usage</span></a> section for a full list):</p>
<table border="1" class="docutils">
<colgroup>
<col width="32%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Option</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">--chrom</span></code></td>
<td>The chromosome to impute (here, the autosomes).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">--bfile</span></code></td>
<td>The genotypes of the study cohort to be imputed.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">--shapeit-bin</span></code></td>
<td>The <code class="docutils literal notranslate"><span class="pre">shapeit</span></code> binary.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">--impute2-bin</span></code></td>
<td>The <code class="docutils literal notranslate"><span class="pre">impute2</span></code> binary.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">--plink-bin</span></code></td>
<td>The <code class="docutils literal notranslate"><span class="pre">plink</span></code> binary.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">--reference</span></code></td>
<td>The <em>fasta</em> file containing the reference genome
for initial strand verification (<em>optional</em>).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">--hap-template</span></code></td>
<td>The template for <em>IMPUTE2</em>’s reference haplotype
files (<code class="docutils literal notranslate"><span class="pre">{chrom}</span></code> will be replaced by the
chromosome number).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">--legend-template</span></code></td>
<td>The template for <em>IMPUTE2</em>’s reference legend
files (<code class="docutils literal notranslate"><span class="pre">{chrom}</span></code> will be replaced by the
chromosome number).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">--map-template</span></code></td>
<td>The template for <em>IMPUTE2</em>’s reference map
files (<code class="docutils literal notranslate"><span class="pre">{chrom}</span></code> will be replaced by the
chromosome number).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">--sample-file</span></code></td>
<td>The name of <em>IMPUTE2</em>’s reference sample file.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">--filtering-rules</span></code></td>
<td>Rules used by <em>IMPUTE2</em> to exclude sites from
its reference files (using the legend files).
Each terms are joined using a logical <em>OR</em>.
Here, loci with an alternative allele frequency
lower than 0.01 and higher than 0.99 will be
excluded from the imputation reference.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">--thread</span></code></td>
<td>The number of thread to use for the analysis.
When using <em>DRMAA</em>, this will be the number of
simultaneous tasks. Four threads will be used.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">--report-title</span></code></td>
<td>The title of the automatic report.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">--report-number</span></code></td>
<td>The number of the report (will appear as
sub-title and in the footer of the automatic
report).</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>If the pipeline fails (<em>e.g.</em> not enough memory or the walltime exceeded on
a computing cluster), re-running the pipeline (with different number of
thread or different walltime) will only launch the task that were not
completed.</p>
<p class="last">The pipeline checks if output files are missing. If an output file is
deleted, the step producing this file will be run again (but not the
subsequent steps).</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Four options will modify the report content: <code class="docutils literal notranslate"><span class="pre">--report-number</span></code>,
<code class="docutils literal notranslate"><span class="pre">--report-title</span></code>, <code class="docutils literal notranslate"><span class="pre">--report-author</span></code> and <code class="docutils literal notranslate"><span class="pre">--report-background</span></code>.</p>
</div>
</div>
<div class="section" id="compiling-the-report">
<span id="genipe-tut-compile-report"></span><h2><a class="toc-backref" href="#id6">Compiling the report</a><a class="headerlink" href="#compiling-the-report" title="Permalink to this headline">¶</a></h2>
<p>A report containing useful information (such as quality metrics and execution
time, among others) is automatically generated once the imputation process is
completed. To compile the report, perform the following commands:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial/genipe/report

make <span class="o">&amp;&amp;</span> make clean
</pre></div>
</div>
<p>This will generate the following
<a class="reference external" href="http://pgxcentre.github.io/genipe/_static/tutorial/report.pdf">PDF report</a>
(which is named <code class="docutils literal notranslate"><span class="pre">report.pdf</span></code>). It is always possible to modify the original
<code class="docutils literal notranslate"><span class="pre">report.tex</span></code> file to include analysis specific details (<em>e.g.</em> cohort
description).</p>
</div>
<div class="section" id="output-files">
<span id="genipe-tut-output-files"></span><h2><a class="toc-backref" href="#id7">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>All results will be located in the <code class="docutils literal notranslate"><span class="pre">genipe</span></code> directory (or whatever
<code class="docutils literal notranslate"><span class="pre">--output-dir</span></code> links to). Here is the directory tree summarizing the output
files.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>genipe/
│
├── chr1/
│   ├── chr1.1_5000000.impute2
│   ├── chr1.1_5000000.impute2_info
│   ├── chr1.1_5000000.impute2_info_by_sample
│   ├── chr1.1_5000000.impute2_summary
│   ├── chr1.1_5000000.impute2_warnings
│   ├── ...
│   ├── chr1.final.bed
│   ├── chr1.final.bim
│   ├── chr1.final.fam
│   ├── chr1.final.log
│   ├── chr1.final.phased.haps
│   ├── chr1.final.phased.ind.me
│   ├── chr1.final.phased.ind.mm
│   ├── chr1.final.phased.log
│   ├── chr1.final.phased.sample
│   ├── chr1.final.phased.snp.me
│   ├── chr1.final.phased.snp.mm
│   ├── ...
│   │
│   └── final_impute2/
│       ├── chr1.imputed.alleles
│       ├── chr1.imputed.completion_rates
│       ├── chr1.imputed.good_sites
│       ├── chr1.imputed.impute2.gz
│       ├── chr1.imputed.impute2_info
│       ├── chr1.imputed.imputed_sites
│       ├── chr1.imputed.log
│       ├── chr1.imputed.maf
│       ├── chr1.imputed.map
│       └── chr1.imputed.sample
│
├── .../
│
├── chromosome_lengths.txt
├── exclusion_summary.txt
├── frequency_pie.pdf
├── genipe.log
├── markers_to_exclude.txt
├── markers_to_flip.txt
│
├── missing
│   ├── missing.imiss
│   ├── missing.lmiss
│   └── missing.log
│
├── report
│   ├── frequency_pie.pdf
│   ├── Makefile
│   ├── references.bib
│   ├── references.bst
│   └── report.tex
│
└── tasks.db
</pre></div>
</div>
<div class="section" id="genipe-directory">
<h3><code class="docutils literal notranslate"><span class="pre">genipe</span></code> directory<a class="headerlink" href="#genipe-directory" title="Permalink to this headline">¶</a></h3>
<p>This directory contains all the chromosome specific analysis. The specific
directory content is describe below. The following files are created inside the
<code class="docutils literal notranslate"><span class="pre">genipe</span></code> directory:</p>
<table border="1" class="docutils">
<colgroup>
<col width="39%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">File</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">chromosome_lengths.txt</span></code></td>
<td>The length of each chromosome (this
information is fetched from Ensembl using
its REST API and saved to file).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">exclusion_summary.txt</span></code></td>
<td>A file containing the exclusion summary
prior to phasing (<em>e.g</em> the number of
ambiguous markers, the number of flipped
markers, etc.).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">frequency_pie.pdf</span></code></td>
<td>This file contains a pie chart describing
the minor allele frequency distribution of
the imputed markers. This file is generated
only if the <code class="xref py py-mod docutils literal notranslate"><span class="pre">matplotlib</span></code> module is
installed.</td>
</tr>
<tr class="row-odd"><td colspan="2"><code class="docutils literal notranslate"><span class="pre">genipe.log</span></code>               | The log file of the main pipeline.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">markers_to_exclude.txt</span></code></td>
<td>The list of markers to exclude prior to
phasing.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">markers_to_flip.txt</span></code></td>
<td>The list of markers to flip prior to
phasing.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">tasks.db</span></code></td>
<td>The <em>sqlite</em> database containing
information of all tasks (if it’s
completed, execution time, etc).</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="genipe-chrn-directories">
<h3><code class="docutils literal notranslate"><span class="pre">genipe/chrN</span></code> directories<a class="headerlink" href="#genipe-chrn-directories" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">chrN</span></code> directories contain the intermediate files, created throughout the
pipeline. The most important files in these directories are the log files (for
errors and summary statistics). There will be one directory per autosomal
chromosomes.</p>
</div>
<div class="section" id="genipe-chrn-final-impute2-directories">
<span id="genipe-tut-output-files-final-impute2"></span><h3><code class="docutils literal notranslate"><span class="pre">genipe/chrN/final_impute2</span></code> directories<a class="headerlink" href="#genipe-chrn-final-impute2-directories" title="Permalink to this headline">¶</a></h3>
<p>These <code class="docutils literal notranslate"><span class="pre">final_impute2</span></code> directories (located in the <code class="docutils literal notranslate"><span class="pre">genipe/chrN</span></code>
directories) contain the final output files from the pipeline for each
autosomal chromosomes. They will contain the following files:</p>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="57%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Extension</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">.imputed.alleles</span></code></td>
<td>Description of the reference and
alternative allele at each sites.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">.imputed.completion_rates</span></code></td>
<td>Number of missing values and completion
rate for all sites (using the
probability threshold set by the user,
where the default is higher and equal
to 0.9).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">.imputed.good_sites</span></code></td>
<td>List of sites which pass the completion
rate threshold (set by the user, where
the default is higher and equal to 0.98)
using the probability threshold (set by
the user, where the default is higher
and equal to 0.9). The sites also pass
the INFO threshold (set by the user,
where the default is higher and equal to
0).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">.imputed.impute2</span></code> or
<code class="docutils literal notranslate"><span class="pre">.imputed.impute2.gz</span></code></td>
<td>Imputation results (merged from the
individual segment files. This file
might be compress (with the <code class="docutils literal notranslate"><span class="pre">.gz</span></code>
extension) if the <code class="docutils literal notranslate"><span class="pre">--bgzip</span></code> option was
used when launching the pipeline.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">.imputed.impute2_info</span></code></td>
<td>Marker-wise information file with one
line per marker and a single header line
at the begening. It contains, among
others, the information value which is a
measure of the observed statistical
information associated with the allele
frequency estimate.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">.imputed.imputed_sites</span></code></td>
<td>List of imputed sites (excluding sites
that were previously genotyped in the
study cohort).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">.imputed.log</span></code></td>
<td>The log file of the merging step.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">.imputed.maf</span></code></td>
<td>File containing the minor allele
frequency (along with minor allele
identification) for all sites using the
probabilitty threshold of 0.9. When no
genotypes are available (because they
are all below the threshold), the MAF is
<code class="docutils literal notranslate"><span class="pre">NA</span></code>.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal notranslate"><span class="pre">.imputed.map</span></code></td>
<td>A <em>map</em> file describing the genomic
location of all sites.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal notranslate"><span class="pre">.imputed.sample</span></code></td>
<td>The sample file generated by the phasing
step, which describe the sample ordering
in the IMPUTE2 files.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="genipe-missing-directory">
<h3><code class="docutils literal notranslate"><span class="pre">genipe/missing</span></code> directory<a class="headerlink" href="#genipe-missing-directory" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">missing</span></code> directory contains the missing rates for both samples
(<code class="docutils literal notranslate"><span class="pre">missing.imiss</span></code>) and genotypes markers (<code class="docutils literal notranslate"><span class="pre">missing.lmiss</span></code>). Those files are
generated by Plink.</p>
</div>
<div class="section" id="genipe-report-directory">
<h3><code class="docutils literal notranslate"><span class="pre">genipe/report</span></code> directory<a class="headerlink" href="#genipe-report-directory" title="Permalink to this headline">¶</a></h3>
<p>This <code class="docutils literal notranslate"><span class="pre">report</span></code> directory contains the automatically generated report, which
provides valuable information about the imputation analysis. Such information
contains cross-validation statistics (as provided by IMPUTE2), frequency
statistics and completion rates according to user defined parameters.</p>
<p>The automatic report is generated in the <code class="docutils literal notranslate"><span class="pre">LaTeX</span></code> language (file
<code class="docutils literal notranslate"><span class="pre">report.tex</span></code>), and can be compile using the following command (as long as
<code class="docutils literal notranslate"><span class="pre">LaTeX</span></code> is installed).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial/genipe/report
make <span class="o">&amp;&amp;</span> make clean
</pre></div>
</div>
<p>This will generate the following
<a class="reference external" href="http://pgxcentre.github.io/genipe/_static/tutorial/report.pdf">PDF report</a>
(which is named <code class="docutils literal notranslate"><span class="pre">report.pdf</span></code>). It is always possible to modify the original
<code class="docutils literal notranslate"><span class="pre">report.tex</span></code> file to include analysis specific details (<em>e.g.</em> cohort
description).</p>
</div>
</div>
<div class="section" id="using-genipe-on-a-server">
<span id="genipe-tut-drmaa"></span><h2><a class="toc-backref" href="#id8">Using genipe on a server</a><a class="headerlink" href="#using-genipe-on-a-server" title="Permalink to this headline">¶</a></h2>
<div class="section" id="drmaa-configuration">
<h3>DRMAA configuration<a class="headerlink" href="#drmaa-configuration" title="Permalink to this headline">¶</a></h3>
<p>If the pipeline is to be launch on a computing server, the <code class="docutils literal notranslate"><span class="pre">--use-drmaa</span></code>
option should be used. This will launch each step on the server using the DRMAA
api. On some cluster, supplemental information is required for each task
(<em>i.e.</em> execution time, number of nodes/processes to reserve). This
parametrization is done using a configuration (<em>ini</em>) file, describing these
parameters for each step.</p>
<p>When providing an empty <em>ini</em> file, the default walltime and number of
nodes/processes will be 15 minutes and 1/1, respectively. Otherwise, different
parameters can be used for each step. For example, the following configuration
will increase the walltime for all phasing tasks from 15 minutes to 3 hours. It
will also run each phasing tasks on one node using 12 processes.</p>
<div class="highlight-ini notranslate"><div class="highlight"><pre><span></span><span class="k">[shapeit_phase]</span>
<span class="na">walltime</span> <span class="o">=</span> <span class="s">03:00:00</span>
<span class="na">nodes</span>    <span class="o">=</span> <span class="s">1</span>
<span class="na">ppn</span>      <span class="o">=</span> <span class="s">12</span>
</pre></div>
</div>
<p>The following example has the same configuration as the previous example, but
will increase the walltime for chromosome 2 to 4 hours, with 1 node and 24
processes.</p>
<div class="highlight-ini notranslate"><div class="highlight"><pre><span></span><span class="k">[shapeit_phase]</span>
<span class="na">walltime</span> <span class="o">=</span> <span class="s">03:00:00</span>
<span class="na">nodes</span>    <span class="o">=</span> <span class="s">1</span>
<span class="na">ppn</span>      <span class="o">=</span> <span class="s">12</span>

<span class="na">chr2_walltime</span> <span class="o">=</span> <span class="s">04:00:00</span>
<span class="na">chr2_nodes</span>    <span class="o">=</span> <span class="s">1</span>
<span class="na">chr2_ppn</span>      <span class="o">=</span> <span class="s">24</span>
</pre></div>
</div>
<p>Since imputation is performed on segments for each chromosome, it is possible
to modify the parameters for a single segment. This is usefull when a segment
doesn’t have time to finish and its imputation requires a rerun. For example,
the following parameters will increase the walltime from 15 minutes to 3.5
hours for segment 10,000,001-15,000,000 on chromosome 1. Also, all segments
located on chromosome 2 will have a walltime of 4 hours.</p>
<div class="highlight-ini notranslate"><div class="highlight"><pre><span></span><span class="k">[impute2]</span>
<span class="na">chr1_10000001_15000000_walltime</span> <span class="o">=</span> <span class="s">03:30:00</span>

<span class="na">chr2_walltime</span> <span class="o">=</span> <span class="s">04:00:00</span>
</pre></div>
</div>
<p>We provide a
<a class="reference external" href="http://pgxcentre.github.io/genipe/_static/tutorial/config_example.ini">configuration example</a>
including all possible section. Also, here is a list of all possible section
(<em>i.e</em> pipeline step) that can be parametrized.</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">plink_exclude</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">plink_missing_rate</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">shapeit_check_1</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">plink_flip</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">shapeit_check_2</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">plink_final_exclude</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">shapeit_phase</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">impute2</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">merge_impute2</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">bgzip</span></code></li>
</ul>
<p>Some cluster doesn’t require any configuration at all. To skip configuration,
use the <code class="docutils literal notranslate"><span class="pre">main</span></code> section of the <em>ini</em> file as such:</p>
<div class="highlight-ini notranslate"><div class="highlight"><pre><span></span><span class="k">[main]</span>
<span class="na">skip_drmaa_config</span> <span class="o">=</span> <span class="s">yes</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Keep in mind that lines starting with a <code class="docutils literal notranslate"><span class="pre">#</span></code> are comments and are not used
in the DRMAA configuration. This is useful to describe what parameters are
used for each step.</p>
</div>
</div>
<div class="section" id="drmaa-preamble">
<h3>DRMAA preamble<a class="headerlink" href="#drmaa-preamble" title="Permalink to this headline">¶</a></h3>
<p>When using the <code class="docutils literal notranslate"><span class="pre">--use-drmaa</span></code> option, the pipeline creates <em>bash</em> script that
are launched on the computing cluster. Some clusters require module to be
loaded and the python virtual environment to be loaded before executing a
script. This is done using the preamble file (the <code class="docutils literal notranslate"><span class="pre">--preamble</span></code> option).</p>
<p>The content of the file will be added between the first line of the temporary
<em>bash</em> script (the <em>shebang</em>) and the actual command. For example, the
following file will load the gcc module (version 4.8.2) and the python virtual
environment before launching the task.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Loading the required module</span>
module load gcc/4.8.2

<span class="c1"># The python virtual environment</span>
<span class="nb">source</span> <span class="nv">$HOME</span>/softwares/python_env/bin/activate
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The preamble file is system dependent, but you should always at least
activate the virtual python environment so that the tools provided by
<a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genipe</span></code></a> are automatically in the system path.</p>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The preamble will be added <strong>as-is</strong> in the <em>bash</em> script that will be
executed. Hence, always be careful of what is included in the preamble.</p>
</div>
</div>
</div>
<div class="section" id="data-in-more-details">
<span id="genipe-tut-more-details"></span><h2><a class="toc-backref" href="#id9">Data in more details</a><a class="headerlink" href="#data-in-more-details" title="Permalink to this headline">¶</a></h2>
<p>First, we will create a project directory where all the analysis will be
performed:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir -p <span class="nv">$HOME</span>/genipe_tutorial
<span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial
</pre></div>
</div>
<div class="section" id="required-softwares">
<span id="genipe-tut-softwares"></span><h3>Required softwares<a class="headerlink" href="#required-softwares" title="Permalink to this headline">¶</a></h3>
<p>The main <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genipe</span></code></a> pipeline requires three external tools:
<a class="reference external" href="http://pngu.mgh.harvard.edu/~purcell/plink/">Plink</a>,
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html">IMPUTE2</a> and
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html">SHAPEIT</a>.
These tools are not required to be located in your <code class="docutils literal notranslate"><span class="pre">PATH</span></code> variable, since you
can specify each of their location at runtime.</p>
<p>If some of the binaries are missing in you <code class="docutils literal notranslate"><span class="pre">PATH</span></code> variable, we will create a
directory where the missing binaries will go:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir -p <span class="nv">$HOME</span>/genipe_tutorial/bin
<span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial/bin
</pre></div>
</div>
<p>Plink is freely available. Go to
<a class="reference external" href="http://pngu.mgh.harvard.edu/~purcell/plink/download.shtml">Plink download page</a>,
read the license, and download the most recent version. Extract the binary and
copy it to the <code class="docutils literal notranslate"><span class="pre">bin</span></code> directory previously created.</p>
<p>IMPUTE2 is freely available for academic use. Go to
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html#download">IMPUTE2 download page</a>,
read the license, and download the most recent version. Extract the binary and
copy it to the <code class="docutils literal notranslate"><span class="pre">bin</span></code> directory previously created.</p>
<p>SHAPEIT is freely available for academic use. Go to
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#download">SHAPEIT download page</a>,
read the license, and download the most recent version. Extract the binary and
copy it to the <code class="docutils literal notranslate"><span class="pre">bin</span></code> directory previously created.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">bin</span></code> directory should now contain the missing binaries: <code class="docutils literal notranslate"><span class="pre">impute2</span></code>,
<code class="docutils literal notranslate"><span class="pre">plink</span></code> and/or <code class="docutils literal notranslate"><span class="pre">shapeit</span></code>.</p>
</div>
<div class="section" id="input-files">
<span id="genipe-tut-input-files"></span><h3>Input files<a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h3>
<div class="section" id="genotypes">
<h4>Genotypes<a class="headerlink" href="#genotypes" title="Permalink to this headline">¶</a></h4>
<p>The genotypes of the study cohort have to be in Plink’s binary pedfile format.
This format allows for time and disk space optimization. Each dataset is
comprised of three files: the <code class="docutils literal notranslate"><span class="pre">.bed</span></code> (binary file, genotype information),
<code class="docutils literal notranslate"><span class="pre">.fam</span></code> (text file, sample information) and <code class="docutils literal notranslate"><span class="pre">.bim</span></code> (text file, marker
information) files. Those file are generated by Plink.</p>
<p>Genotype data are provided for this tutorial. They are a subset from the files
provided by Plink on its
<a class="reference external" href="http://pngu.mgh.harvard.edu/~purcell/plink/res.shtml">resources page</a>.
The build was lifted over from <em>GRCh36</em> to <em>GRCh37</em> using
<a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgLiftOver">USCS liftOver</a>. Only CEU samples
were kept, along with markers having a completion rate higher or equal to 0.98.
To download the data for this tutorial, execute the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir -p <span class="nv">$HOME</span>/genipe_tutorial/data
<span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial/data

wget http://pgxcentre.github.io/genipe/_static/tutorial/hapmap_CEU_r23a_hg19.tar.bz2
tar -jxf hapmap_CEU_r23a_hg19.tar.bz2
rm hapmap_CEU_r23a_hg19.tar.bz2
</pre></div>
</div>
</div>
<div class="section" id="reference-panels">
<h4>Reference panels<a class="headerlink" href="#reference-panels" title="Permalink to this headline">¶</a></h4>
<p>IMPUTE2 can use publicly available reference datasets. They provide such
dataset on their website. Go to IMPUTE2’s
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html#reference">reference page</a>,
and download the most recent reference data (which is over 12Gb). Once the
reference is downloaded, extract it in the working directory
(<code class="docutils literal notranslate"><span class="pre">$HOME/genipe_tutorial</span></code>).</p>
<p>The following commands should download the reference files (1000 Genomes phase
3) and extract them in the required directory. Note that the specified URL
might change.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial

wget https://mathgen.stats.ox.ac.uk/impute/1000GP_Phase3.tgz
tar -zxf 1000GP_Phase3.tgz
rm 1000GP_Phase3.tgz
</pre></div>
</div>
</div>
<div class="section" id="human-reference-optional">
<h4>Human reference (optional)<a class="headerlink" href="#human-reference-optional" title="Permalink to this headline">¶</a></h4>
<p>The pipeline include an optional step to check for strand alignment with the
reference panel (using <em>SHAPEIT</em>). The drawback of this method is that it is
impossible to verify the strand of markers which are absent from the
<em>IMPUTE2</em>’s reference. We have introduce a way to check the strand using the
reference genome (in <em>fasta</em> format, indexed using <em>faidx</em>).</p>
<p>We have created such reference using the
<a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/">UCSC’s human reference</a>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir -p <span class="nv">$HOME</span>/genipe_tutorial/hg19
<span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial/hg19

wget http://statgen.org/wp-content/uploads/Softwares/genipe/supp_files/hg19.tar.bz2
tar -jxf hg19.tar.bz2
rm hg19.tar.bz2
</pre></div>
</div>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_mixedlm.html" title="Linear mixed effects model"
             >next</a> |</li>
        <li class="right" >
          <a href="../tutorials.html" title="Short Tutorials"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">genipe 1.4.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorials.html" >Short Tutorials</a> &#187;</li> 
      </ul>
    </div>


    <div class="footer" role="contentinfo">
        &#169; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.7.5.
    </div>

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-81009922-1', 'auto');
  ga('send', 'pageview');

</script>


  </body>
</html>