

<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>genipe.tools.impute2_extractor &#8212; genipe 1.4.2 documentation</title>
    <link rel="stylesheet" href="../../../_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../../_static/bizstyle.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">genipe 1.4.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../../genipe.html" accesskey="U">genipe</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for genipe.tools.impute2_extractor</h1><div class="highlight"><pre>
<span></span>
<span class="c1"># This file is part of genipe.</span>
<span class="c1">#</span>
<span class="c1"># This work is licensed under the Creative Commons Attribution-NonCommercial</span>
<span class="c1"># 4.0 International License. To view a copy of this license, visit</span>
<span class="c1"># http://creativecommons.org/licenses/by-nc/4.0/ or send a letter to Creative</span>
<span class="c1"># Commons, PO Box 1866, Mountain View, CA 94042, USA.</span>


<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">shlex</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">namedtuple</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">numpy</span> <span class="k">import</span> <span class="n">nan</span>

<span class="kn">from</span> <span class="nn">..formats</span> <span class="k">import</span> <span class="n">index</span>
<span class="kn">from</span> <span class="nn">..formats</span> <span class="k">import</span> <span class="n">impute2</span>
<span class="kn">from</span> <span class="nn">..error</span> <span class="k">import</span> <span class="n">GenipeError</span>
<span class="kn">from</span> <span class="nn">..</span> <span class="k">import</span> <span class="n">__version__</span><span class="p">,</span> <span class="n">chromosomes</span>

<span class="c1"># Check if pyplink is installed</span>
<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">pyplink</span> <span class="k">import</span> <span class="n">PyPlink</span>
    <span class="n">HAS_PYPLINK</span> <span class="o">=</span> <span class="kc">True</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="n">HAS_PYPLINK</span> <span class="o">=</span> <span class="kc">False</span>


<span class="n">__author__</span> <span class="o">=</span> <span class="s2">&quot;Louis-Philippe Lemieux Perreault&quot;</span>
<span class="n">__copyright__</span> <span class="o">=</span> <span class="s2">&quot;Copyright 2014, Beaulieu-Saucier Pharmacogenomics Centre&quot;</span>
<span class="n">__license__</span> <span class="o">=</span> <span class="s2">&quot;Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)&quot;</span>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_extractor.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;The main function.</span>

<span class="sd">    Args:</span>
<span class="sd">        args (argparse.Namespace): the arguments to be parsed (if</span>
<span class="sd">                                   :py:func:`main` is called by another</span>
<span class="sd">                                   modulel)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Creating the option parser</span>
    <span class="n">desc</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Extract imputed markers located in a specific genomic region. &quot;</span>
            <span class="s2">&quot;This script is part of the &#39;genipe&#39; package, version &quot;</span>
            <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">))</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="n">desc</span><span class="p">)</span>

    <span class="c1"># Files that need closing</span>
    <span class="n">logging_fh</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Parsing the options</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parse_args</span><span class="p">(</span><span class="n">parser</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

        <span class="c1"># The logging handlers</span>
        <span class="n">handlers</span> <span class="o">=</span> <span class="p">[</span><span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">()]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">index_only</span><span class="p">:</span>
            <span class="n">log_file</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.log&quot;</span>
            <span class="n">logging_fh</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">FileHandler</span><span class="p">(</span><span class="n">log_file</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">)</span>
            <span class="n">handlers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">logging_fh</span><span class="p">)</span>

        <span class="c1"># Adding the logging capability</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
            <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;[</span><span class="si">%(asctime)s</span><span class="s2"> </span><span class="si">%(levelname)s</span><span class="s2">] </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">datefmt</span><span class="o">=</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">,</span>
            <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">debug</span> <span class="k">else</span> <span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span>
            <span class="n">handlers</span><span class="o">=</span><span class="n">handlers</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># First log</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">index_only</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Logging everything into &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">log_file</span><span class="p">))</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Program arguments: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">shlex</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">part</span><span class="p">)</span> <span class="k">for</span> <span class="n">part</span> <span class="ow">in</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
        <span class="p">))</span>

        <span class="c1"># Checking the options</span>
        <span class="n">check_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">index_only</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">index_file</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">impute2</span><span class="p">)</span>

        <span class="c1"># Gathering what needs to be extracted</span>
        <span class="n">to_extract</span> <span class="o">=</span> <span class="n">gather_extraction</span><span class="p">(</span>
            <span class="n">fn</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">impute2</span><span class="p">,</span>
            <span class="n">maf</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">maf</span><span class="p">,</span>
            <span class="n">rate</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">rate</span><span class="p">,</span>
            <span class="n">info</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">info</span><span class="p">,</span>
            <span class="n">extract_filename</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">extract</span><span class="p">,</span>
            <span class="n">genomic_range</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">genomic</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># Extraction</span>
        <span class="n">extract_markers</span><span class="p">(</span>
            <span class="n">fn</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">impute2</span><span class="p">,</span>
            <span class="n">to_extract</span><span class="o">=</span><span class="n">to_extract</span><span class="p">,</span>
            <span class="n">out_prefix</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">,</span>
            <span class="n">out_format</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out_format</span><span class="p">,</span>
            <span class="n">prob_t</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">prob</span><span class="p">,</span>
            <span class="n">is_long</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">long_format</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Catching the Ctrl^C</span>
    <span class="k">except</span> <span class="ne">KeyboardInterrupt</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Cancelled by user&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Catching the GenipeError</span>
    <span class="k">except</span> <span class="n">GenipeError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="k">raise</span>

    <span class="k">finally</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">logging_fh</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">logging_fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>


<div class="viewcode-block" id="index_file"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_extractor.index_file">[docs]</a><span class="k">def</span> <span class="nf">index_file</span><span class="p">(</span><span class="n">fn</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Indexes the impute2 file.</span>

<span class="sd">    Args:</span>
<span class="sd">        fn (str): the name of the impute2 file</span>

<span class="sd">    This function uses the :py:func:`genipe.formats.index.get_index` to create</span>
<span class="sd">    the index file if it&#39;s missing.</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">        We won&#39;t catch the :py:class:`genipe.error.GenipeError` exception if</span>
<span class="sd">        it&#39;s raised, since the message will be relevant to the user.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># For each input file</span>
    <span class="n">index</span><span class="o">.</span><span class="n">get_index</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">cols</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">],</span>
                    <span class="n">sep</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="extract_markers"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_extractor.extract_markers">[docs]</a><span class="k">def</span> <span class="nf">extract_markers</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">to_extract</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">,</span> <span class="n">out_format</span><span class="p">,</span> <span class="n">prob_t</span><span class="p">,</span> <span class="n">is_long</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extracts according to names.</span>

<span class="sd">    Args:</span>
<span class="sd">        fn (str): the name of the input file</span>
<span class="sd">        to_extract (set): the list of markers to extract for each input file</span>
<span class="sd">        out_prefix (str): the output prefix</span>
<span class="sd">        out_format (list): the output format(s)</span>
<span class="sd">        prob_t (float): the probability threshold</span>
<span class="sd">        is_long (bool): True if format needs to be long</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># The output files (probabilities)</span>
    <span class="n">o_files</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">suffix</span><span class="p">:</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.&quot;</span> <span class="o">+</span> <span class="n">suffix</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">suffix</span> <span class="ow">in</span> <span class="n">out_format</span> <span class="k">if</span> <span class="n">suffix</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;bed&quot;</span><span class="p">}</span>
    <span class="p">}</span>

    <span class="c1"># If there is the &#39;bed&#39; format, we actually need pyplink</span>
    <span class="k">if</span> <span class="s2">&quot;bed&quot;</span> <span class="ow">in</span> <span class="n">out_format</span><span class="p">:</span>
        <span class="n">o_files</span><span class="p">[</span><span class="s2">&quot;bed&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">PyPlink</span><span class="p">(</span><span class="n">out_prefix</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">),</span>
            <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.bim&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">),</span>
        <span class="p">)</span>

    <span class="c1"># Creating a fam (if bed)</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="n">get_samples</span><span class="p">(</span><span class="n">get_file_prefix</span><span class="p">(</span><span class="n">fn</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.sample&quot;</span><span class="p">)</span>
    <span class="n">sample_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">/</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">id_1</span><span class="p">,</span> <span class="n">id_2</span><span class="p">)</span> <span class="k">for</span> <span class="n">id_1</span><span class="p">,</span> <span class="n">id_2</span>
                    <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">samples</span><span class="o">.</span><span class="n">ID_1</span><span class="p">,</span> <span class="n">samples</span><span class="o">.</span><span class="n">ID_2</span><span class="p">)]</span>

    <span class="c1"># Writing the header (for dosage)</span>
    <span class="k">if</span> <span class="s2">&quot;dosage&quot;</span> <span class="ow">in</span> <span class="n">o_files</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">is_long</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;fid&quot;</span><span class="p">,</span> <span class="s2">&quot;iid&quot;</span><span class="p">,</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;minor&quot;</span><span class="p">,</span> <span class="s2">&quot;major&quot;</span><span class="p">,</span>
                  <span class="s2">&quot;dosage&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">o_files</span><span class="p">[</span><span class="s2">&quot;dosage&quot;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;minor&quot;</span><span class="p">,</span> <span class="s2">&quot;major&quot;</span><span class="p">,</span>
                  <span class="o">*</span><span class="n">sample_names</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">o_files</span><span class="p">[</span><span class="s2">&quot;dosage&quot;</span><span class="p">])</span>

    <span class="c1"># Writing the header (for calls)</span>
    <span class="k">if</span> <span class="s2">&quot;calls&quot;</span> <span class="ow">in</span> <span class="n">o_files</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">is_long</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;fid&quot;</span><span class="p">,</span> <span class="s2">&quot;iid&quot;</span><span class="p">,</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;cm&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="s2">&quot;call&quot;</span><span class="p">,</span>
                  <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">o_files</span><span class="p">[</span><span class="s2">&quot;calls&quot;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;cm&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span>
                  <span class="o">*</span><span class="n">sample_names</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">o_files</span><span class="p">[</span><span class="s2">&quot;calls&quot;</span><span class="p">])</span>

    <span class="c1"># Extracted positions</span>
    <span class="n">all_extracted</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

    <span class="c1"># Reading the impute2 file</span>
    <span class="n">extracted</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

    <span class="c1"># Finding the name of the file containing the index</span>
    <span class="n">file_index</span> <span class="o">=</span> <span class="n">index</span><span class="o">.</span><span class="n">get_index</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">cols</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
                                 <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">)</span>

    <span class="c1"># Keeping only required values from the index</span>
    <span class="n">file_index</span> <span class="o">=</span> <span class="n">file_index</span><span class="p">[</span><span class="n">file_index</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">to_extract</span><span class="p">)]</span>

    <span class="c1"># Getting all the markers value</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Extracting </span><span class="si">{:,d}</span><span class="s2"> markers&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">file_index</span><span class="p">)))</span>
    <span class="k">with</span> <span class="n">index</span><span class="o">.</span><span class="n">get_open_func</span><span class="p">(</span><span class="n">fn</span><span class="p">)(</span><span class="n">fn</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">seek_value</span> <span class="ow">in</span> <span class="n">file_index</span><span class="o">.</span><span class="n">seek</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
            <span class="c1"># Seeking</span>
            <span class="n">i_file</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">seek_value</span><span class="p">))</span>

            <span class="c1"># Reading the line</span>
            <span class="n">line</span> <span class="o">=</span> <span class="n">i_file</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span>

            <span class="c1"># The marker name</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

            <span class="c1"># Printing the data</span>
            <span class="n">print_data</span><span class="p">(</span><span class="n">o_files</span><span class="p">,</span> <span class="n">prob_t</span><span class="p">,</span> <span class="n">samples</span><span class="o">.</span><span class="n">ID_1</span><span class="p">,</span> <span class="n">samples</span><span class="o">.</span><span class="n">ID_2</span><span class="p">,</span> <span class="n">line</span><span class="o">=</span><span class="n">line</span><span class="p">,</span>
                       <span class="n">row</span><span class="o">=</span><span class="n">row</span><span class="p">,</span> <span class="n">is_long</span><span class="o">=</span><span class="n">is_long</span><span class="p">)</span>

            <span class="c1"># Saving statistics</span>
            <span class="n">extracted</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Extracted </span><span class="si">{:,d}</span><span class="s2"> markers&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">extracted</span><span class="p">)))</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">to_extract</span> <span class="o">-</span> <span class="n">extracted</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Missing </span><span class="si">{:,d}</span><span class="s2"> &quot;</span>
                        <span class="s2">&quot;markers&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">to_extract</span> <span class="o">-</span> <span class="n">extracted</span><span class="p">)))</span>

    <span class="c1"># Keeping track of what has been extracted</span>
    <span class="n">all_extracted</span> <span class="o">|=</span> <span class="n">extracted</span>

    <span class="c1"># Extracting the companion files (if impute2 and files are present)</span>
    <span class="k">if</span> <span class="s2">&quot;impute2&quot;</span> <span class="ow">in</span> <span class="n">o_files</span><span class="p">:</span>
        <span class="n">extract_companion_files</span><span class="p">(</span>
            <span class="n">i_prefix</span><span class="o">=</span><span class="n">get_file_prefix</span><span class="p">(</span><span class="n">fn</span><span class="p">),</span>
            <span class="n">to_extract</span><span class="o">=</span><span class="n">to_extract</span><span class="p">,</span>
            <span class="n">o_prefix</span><span class="o">=</span><span class="n">out_prefix</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Writing the FAM file if bed</span>
    <span class="k">if</span> <span class="s2">&quot;bed&quot;</span> <span class="ow">in</span> <span class="n">o_files</span><span class="p">:</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ID_1&quot;</span><span class="p">,</span> <span class="s2">&quot;ID_2&quot;</span><span class="p">,</span> <span class="s2">&quot;father&quot;</span><span class="p">,</span> <span class="s2">&quot;mother&quot;</span><span class="p">,</span> <span class="s2">&quot;sex&quot;</span><span class="p">,</span> <span class="s2">&quot;plink_pheno&quot;</span><span class="p">]</span>
        <span class="n">samples</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.fam&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                             <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Closing the files</span>
    <span class="k">for</span> <span class="n">o_format</span><span class="p">,</span> <span class="n">o_file</span> <span class="ow">in</span> <span class="n">o_files</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">o_format</span> <span class="o">==</span> <span class="s2">&quot;bed&quot;</span><span class="p">:</span>
            <span class="n">o_file</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="n">o_file</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">o_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Extraction complete</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Extraction of </span><span class="si">{:,d}</span><span class="s2"> markers &quot;</span>
                 <span class="s2">&quot;completed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_extracted</span><span class="p">)))</span></div>


<div class="viewcode-block" id="get_samples"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_extractor.get_samples">[docs]</a><span class="k">def</span> <span class="nf">get_samples</span><span class="p">(</span><span class="n">fn</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reads the sample files, and extract the information.</span>

<span class="sd">    Args:</span>
<span class="sd">        fn (str): the name of the sample file</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: the sample information</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Reading the file</span>
    <span class="n">sample</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">)</span>

    <span class="c1"># Removing the first data row</span>
    <span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">1</span><span class="p">:,</span> <span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">sample</span></div>


<div class="viewcode-block" id="extract_companion_files"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_extractor.extract_companion_files">[docs]</a><span class="k">def</span> <span class="nf">extract_companion_files</span><span class="p">(</span><span class="n">i_prefix</span><span class="p">,</span> <span class="n">o_prefix</span><span class="p">,</span> <span class="n">to_extract</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract markers from companion files (if they exists).</span>

<span class="sd">    Args:</span>
<span class="sd">        i_prefix (str): the prefix of the input file</span>
<span class="sd">        o_prefix (str): the prefix of the output file</span>
<span class="sd">        to_extract (set): the set of markers to extract</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">file_info</span> <span class="o">=</span> <span class="p">[</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.alleles&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;name&quot;</span><span class="p">),</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.completion_rates&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;name&quot;</span><span class="p">),</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.good_sites&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.impute2_info&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;name&quot;</span><span class="p">),</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.imputed_sites&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.maf&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;name&quot;</span><span class="p">),</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.map&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
    <span class="p">]</span>

    <span class="k">for</span> <span class="n">info</span> <span class="ow">in</span> <span class="n">file_info</span><span class="p">:</span>
        <span class="c1"># The name of the input file</span>
        <span class="n">i_fn</span> <span class="o">=</span> <span class="n">i_prefix</span> <span class="o">+</span> <span class="n">info</span><span class="p">[</span><span class="s2">&quot;suffix&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">i_fn</span><span class="p">):</span>
            <span class="c1"># The file doesn&#39;t exist, so we continue</span>
            <span class="k">continue</span>

        <span class="c1"># The name of the output file</span>
        <span class="n">o_fn</span> <span class="o">=</span> <span class="n">o_prefix</span> <span class="o">+</span> <span class="n">info</span><span class="p">[</span><span class="s2">&quot;suffix&quot;</span><span class="p">]</span>

        <span class="c1"># If the file doesn&#39;t have a header, we just read line per line</span>
        <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">i_fn</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">o_fn</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o_file</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">i_file</span><span class="p">):</span>
                <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;sep&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span>
                <span class="k">if</span> <span class="n">info</span><span class="p">[</span><span class="s2">&quot;header&quot;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">header</span> <span class="o">=</span> <span class="p">{</span><span class="n">name</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">row</span><span class="p">)}</span>
                    <span class="k">if</span> <span class="n">info</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">header</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: missing column </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="n">i_fn</span><span class="p">,</span>
                            <span class="n">info</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">],</span>
                        <span class="p">))</span>

                    <span class="n">info</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">header</span><span class="p">[</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span>
                    <span class="n">o_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                    <span class="k">continue</span>

                <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]]</span> <span class="ow">in</span> <span class="n">to_extract</span><span class="p">:</span>
                    <span class="n">o_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

    <span class="c1"># We need to copy the sample file</span>
    <span class="n">sample_fn</span> <span class="o">=</span> <span class="n">i_prefix</span> <span class="o">+</span> <span class="s2">&quot;.sample&quot;</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">sample_fn</span><span class="p">):</span>
        <span class="n">o_fn</span> <span class="o">=</span> <span class="n">o_prefix</span> <span class="o">+</span> <span class="s2">&quot;.sample&quot;</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span><span class="n">sample_fn</span><span class="p">,</span> <span class="n">o_fn</span><span class="p">)</span></div>


<div class="viewcode-block" id="print_data"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_extractor.print_data">[docs]</a><span class="k">def</span> <span class="nf">print_data</span><span class="p">(</span><span class="n">o_files</span><span class="p">,</span> <span class="n">prob_t</span><span class="p">,</span> <span class="n">fid</span><span class="p">,</span> <span class="n">iid</span><span class="p">,</span> <span class="n">is_long</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">line</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">row</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Prints an impute2 line.</span>

<span class="sd">    Args:</span>
<span class="sd">        o_files (dict): the output files</span>
<span class="sd">        prob_t (float): the probability threshold</span>
<span class="sd">        fid (list): the list of family IDs</span>
<span class="sd">        iid (list): the list of sample IDs</span>
<span class="sd">        is_long (bool): True if the format is long (dosage, calls)</span>
<span class="sd">        line (str): the impute2 line</span>
<span class="sd">        row (list): the impute2 line, split by spaces</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Probabilities?</span>
    <span class="k">if</span> <span class="s2">&quot;impute2&quot;</span> <span class="ow">in</span> <span class="n">o_files</span><span class="p">:</span>
        <span class="n">o_files</span><span class="p">[</span><span class="s2">&quot;impute2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

    <span class="c1"># Require more?</span>
    <span class="n">a1</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">a2</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">chrom</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">good_calls</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">probabilities</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;dosage&quot;</span> <span class="ow">in</span> <span class="n">o_files</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="s2">&quot;calls&quot;</span> <span class="ow">in</span> <span class="n">o_files</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="s2">&quot;bed&quot;</span> <span class="ow">in</span> <span class="n">o_files</span><span class="p">):</span>
        <span class="c1"># Getting the informations</span>
        <span class="n">marker_info</span><span class="p">,</span> <span class="n">probabilities</span> <span class="o">=</span> <span class="n">impute2</span><span class="o">.</span><span class="n">matrix_from_line</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
        <span class="n">chrom</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span> <span class="o">=</span> <span class="n">marker_info</span>

        <span class="c1"># Getting the good calls</span>
        <span class="n">good_calls</span> <span class="o">=</span> <span class="n">impute2</span><span class="o">.</span><span class="n">get_good_probs</span><span class="p">(</span><span class="n">probabilities</span><span class="p">,</span> <span class="n">min_prob</span><span class="o">=</span><span class="n">prob_t</span><span class="p">)</span>

    <span class="c1"># Dosage?</span>
    <span class="k">if</span> <span class="s2">&quot;dosage&quot;</span> <span class="ow">in</span> <span class="n">o_files</span><span class="p">:</span>
        <span class="c1"># Getting the maf</span>
        <span class="n">maf</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span> <span class="o">=</span> <span class="n">impute2</span><span class="o">.</span><span class="n">maf_from_probs</span><span class="p">(</span>
            <span class="n">prob_matrix</span><span class="o">=</span><span class="n">probabilities</span><span class="p">[</span><span class="n">good_calls</span><span class="p">,</span> <span class="p">:],</span>
            <span class="n">a1</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">a2</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">dosage</span> <span class="o">=</span> <span class="n">impute2</span><span class="o">.</span><span class="n">dosage_from_probs</span><span class="p">(</span>
            <span class="n">homo_probs</span><span class="o">=</span><span class="n">probabilities</span><span class="p">[:,</span> <span class="n">minor</span><span class="p">],</span>
            <span class="n">hetero_probs</span><span class="o">=</span><span class="n">probabilities</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span>
            <span class="n">scale</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">dosage</span><span class="p">[</span><span class="o">~</span><span class="n">good_calls</span><span class="p">]</span> <span class="o">=</span> <span class="n">nan</span>

        <span class="n">alleles</span> <span class="o">=</span> <span class="p">[</span><span class="n">a1</span><span class="p">,</span> <span class="n">nan</span><span class="p">,</span> <span class="n">a2</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">is_long</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">sample_f</span><span class="p">,</span> <span class="n">sample_i</span><span class="p">,</span> <span class="n">sample_d</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">fid</span><span class="p">,</span> <span class="n">iid</span><span class="p">,</span> <span class="n">dosage</span><span class="p">):</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">sample_f</span><span class="p">,</span> <span class="n">sample_i</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">alleles</span><span class="p">[</span><span class="n">minor</span><span class="p">],</span>
                      <span class="n">alleles</span><span class="p">[</span><span class="n">major</span><span class="p">],</span> <span class="n">sample_d</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                      <span class="n">file</span><span class="o">=</span><span class="n">o_files</span><span class="p">[</span><span class="s2">&quot;dosage&quot;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">alleles</span><span class="p">[</span><span class="n">minor</span><span class="p">],</span> <span class="n">alleles</span><span class="p">[</span><span class="n">major</span><span class="p">],</span> <span class="o">*</span><span class="n">dosage</span><span class="p">,</span>
                  <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">o_files</span><span class="p">[</span><span class="s2">&quot;dosage&quot;</span><span class="p">])</span>

    <span class="c1"># Bed?</span>
    <span class="k">if</span> <span class="s2">&quot;bed&quot;</span> <span class="ow">in</span> <span class="n">o_files</span><span class="p">:</span>
        <span class="n">geno</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span> <span class="o">=</span> <span class="n">impute2</span><span class="o">.</span><span class="n">additive_from_probs</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">probabilities</span><span class="p">)</span>
        <span class="n">geno</span><span class="p">[</span><span class="o">~</span><span class="n">good_calls</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
        <span class="n">o_files</span><span class="p">[</span><span class="s2">&quot;bed&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">write_genotypes</span><span class="p">(</span><span class="n">geno</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
              <span class="n">file</span><span class="o">=</span><span class="n">o_files</span><span class="p">[</span><span class="s2">&quot;bed&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>

    <span class="c1"># Hard calls?</span>
    <span class="k">if</span> <span class="s2">&quot;calls&quot;</span> <span class="ow">in</span> <span class="n">o_files</span><span class="p">:</span>
        <span class="n">calls</span> <span class="o">=</span> <span class="n">impute2</span><span class="o">.</span><span class="n">hard_calls_from_probs</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">probabilities</span><span class="p">)</span>
        <span class="n">calls</span><span class="p">[</span><span class="o">~</span><span class="n">good_calls</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;0 0&quot;</span>

        <span class="k">if</span> <span class="n">is_long</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">sample_f</span><span class="p">,</span> <span class="n">sample_i</span><span class="p">,</span> <span class="n">sample_c</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">fid</span><span class="p">,</span> <span class="n">iid</span><span class="p">,</span> <span class="n">calls</span><span class="p">):</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">sample_f</span><span class="p">,</span> <span class="n">sample_i</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">sample_c</span><span class="p">,</span>
                      <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">o_files</span><span class="p">[</span><span class="s2">&quot;calls&quot;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="o">*</span><span class="n">calls</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                  <span class="n">file</span><span class="o">=</span><span class="n">o_files</span><span class="p">[</span><span class="s2">&quot;calls&quot;</span><span class="p">])</span></div>


<div class="viewcode-block" id="gather_extraction"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_extractor.gather_extraction">[docs]</a><span class="k">def</span> <span class="nf">gather_extraction</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">maf</span><span class="p">,</span> <span class="n">rate</span><span class="p">,</span> <span class="n">info</span><span class="p">,</span> <span class="n">extract_filename</span><span class="p">,</span> <span class="n">genomic_range</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Gather positions that are required.</span>

<span class="sd">    Args:</span>
<span class="sd">        fn (str): the impute2 filename</span>
<span class="sd">        maf (float): the minor allele frequency threshold (might be ``None``)</span>
<span class="sd">        rate (float): the call rate threshold (might be ``None``)</span>
<span class="sd">        info (float): the marker information value threshold (might be</span>
<span class="sd">                      ``None``)</span>
<span class="sd">        extract_filename (str): the name of the file containing marker names to</span>
<span class="sd">                                extract (might be ``None``)</span>
<span class="sd">        genomic_range (str): the genomic range for extraction</span>

<span class="sd">    Returns:</span>
<span class="sd">        set: the set of markers to extract</span>

<span class="sd">    If extraction by marker name is required, only those markers will be</span>
<span class="sd">    extracted. Otherwise, ``maf``, ``rate``, ``info`` or ``genomic_range`` can</span>
<span class="sd">    be specified (alone or together) to extract markers according to minor</span>
<span class="sd">    allele frequency, call rate and genomic location.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Gathering information about </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fn</span><span class="p">))</span>

    <span class="c1"># The prefix of all the input files</span>
    <span class="n">prefix</span> <span class="o">=</span> <span class="n">get_file_prefix</span><span class="p">(</span><span class="n">fn</span><span class="p">)</span>

    <span class="c1"># Reading the map file</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading MAP data&quot;</span><span class="p">)</span>
    <span class="n">map_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.map&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span>
                           <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">])</span>
    <span class="n">map_data</span> <span class="o">=</span> <span class="n">map_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="n">verify_integrity</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;MAP data contained </span><span class="si">{:,d}</span><span class="s2"> markers&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">map_data</span><span class="p">)))</span>

    <span class="c1"># If extraction, we only require a list of marker names</span>
    <span class="k">if</span> <span class="n">extract_filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">available_markers</span> <span class="o">=</span> <span class="n">map_data</span><span class="o">.</span><span class="n">index</span>
        <span class="n">marker_list</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">extract_filename</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="n">marker_list</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">i_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">())</span>

        <span class="k">return</span> <span class="nb">set</span><span class="p">(</span><span class="n">available_markers</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">marker_list</span><span class="p">))</span>

    <span class="c1"># Do we require a genomic location?</span>
    <span class="k">if</span> <span class="n">genomic_range</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Keeping markers in required genomic region&quot;</span><span class="p">)</span>
        <span class="n">map_data</span> <span class="o">=</span> <span class="n">map_data</span><span class="p">[(</span>
            <span class="p">(</span><span class="n">map_data</span><span class="o">.</span><span class="n">chrom</span> <span class="o">==</span> <span class="n">genomic_range</span><span class="o">.</span><span class="n">chrom</span><span class="p">)</span> <span class="o">&amp;</span>
            <span class="p">(</span><span class="n">map_data</span><span class="o">.</span><span class="n">pos</span> <span class="o">&gt;=</span> <span class="n">genomic_range</span><span class="o">.</span><span class="n">start</span><span class="p">)</span> <span class="o">&amp;</span>
            <span class="p">(</span><span class="n">map_data</span><span class="o">.</span><span class="n">pos</span> <span class="o">&lt;=</span> <span class="n">genomic_range</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
        <span class="p">)]</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Required genomic region contained </span><span class="si">{:,d}</span><span class="s2"> &quot;</span>
                     <span class="s2">&quot;markers&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">map_data</span><span class="p">)))</span>

    <span class="c1"># Do we require a certain MAF?</span>
    <span class="k">if</span> <span class="n">maf</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading MAF data&quot;</span><span class="p">)</span>
        <span class="n">maf_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.maf&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">maf_data</span> <span class="o">=</span> <span class="n">maf_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="n">verify_integrity</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Merging</span>
        <span class="n">map_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">map_data</span><span class="p">,</span>
            <span class="n">maf_data</span><span class="p">[</span><span class="n">maf_data</span><span class="o">.</span><span class="n">maf</span> <span class="o">&gt;=</span> <span class="n">maf</span><span class="p">],</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
            <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{:,d}</span><span class="s2"> markers with maf &gt;= </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">map_data</span><span class="p">),</span> <span class="n">maf</span><span class="p">))</span>

    <span class="c1"># Do we required a certain completion rate?</span>
    <span class="k">if</span> <span class="n">rate</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading completion rates&quot;</span><span class="p">)</span>
        <span class="n">rate_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.completion_rates&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                                <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
        <span class="n">rate_data</span> <span class="o">=</span> <span class="n">rate_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="n">verify_integrity</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">map_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">map_data</span><span class="p">,</span>
            <span class="n">rate_data</span><span class="p">[</span><span class="n">rate_data</span><span class="o">.</span><span class="n">completion_rate</span> <span class="o">&gt;=</span> <span class="n">rate</span><span class="p">],</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
            <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{:,d}</span><span class="s2"> markers with completion rate &gt;= &quot;</span>
                     <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">map_data</span><span class="p">),</span> <span class="n">rate</span><span class="p">))</span>

    <span class="c1"># Do we required a certain information value?</span>
    <span class="k">if</span> <span class="n">info</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading information values&quot;</span><span class="p">)</span>
        <span class="n">info_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.impute2_info&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">info_data</span> <span class="o">=</span> <span class="n">info_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="n">verify_integrity</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">map_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">map_data</span><span class="p">,</span>
            <span class="n">info_data</span><span class="p">[</span><span class="n">info_data</span><span class="p">[</span><span class="s2">&quot;info&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">info</span><span class="p">],</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
            <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{:,d}</span><span class="s2"> markers with information value &gt;= &quot;</span>
                     <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">map_data</span><span class="p">),</span> <span class="n">info</span><span class="p">))</span>

    <span class="c1"># Extracting the names</span>
    <span class="n">to_extract</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">map_data</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">to_extract</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No marker left for analysis&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">to_extract</span></div>


<div class="viewcode-block" id="get_file_prefix"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_extractor.get_file_prefix">[docs]</a><span class="k">def</span> <span class="nf">get_file_prefix</span><span class="p">(</span><span class="n">fn</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Gets the filename prefix.</span>

<span class="sd">    Args:</span>
<span class="sd">        fn (str): the name of the file from which the prefix is required</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: the prefix of the file</span>

<span class="sd">    This function removes the extension from the file name, and return its</span>
<span class="sd">    prefix (*e.g.* ``test.impute2`` returns ``test``, and</span>
<span class="sd">    ``../test.impute2.gz`` returns ``../test``).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">fn</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">prefix</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;impute2&quot;</span><span class="p">):</span>
        <span class="n">prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">prefix</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">prefix</span></div>


<div class="viewcode-block" id="check_args"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_extractor.check_args">[docs]</a><span class="k">def</span> <span class="nf">check_args</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Checks the arguments and options.</span>

<span class="sd">    Args:</span>
<span class="sd">        args (argparse.Namespace): the options to verify</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">        If there is a problem, a :py:class:`genipe.error.GenipeError` is</span>
<span class="sd">        raised.</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">        Noting is checked (apart from the impute2 files) if indexation is asked</span>
<span class="sd">        (``--index`` option).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Checking that the impute2 file exists</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">impute2</span><span class="p">):</span>
        <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">impute2</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">index_only</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span>

    <span class="c1"># Is there something to extract?</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">genomic</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">maf</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">rate</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">info</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">extract</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;nothing to extract: use &#39;--extract&#39;, &quot;</span>
                              <span class="s2">&quot;&#39;--genomic&#39;, &#39;--maf&#39; or &#39;--rate&#39;&quot;</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">extract</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;&#39;--extract&#39; can only be use alone&quot;</span><span class="p">)</span>

    <span class="c1"># What extensions to look for after checking arguments</span>
    <span class="n">extensions</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

    <span class="c1"># If extract, check the file</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">extract</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">extract</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">extract</span><span class="p">))</span>

    <span class="c1"># If genomic, we check the format</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">genomic</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">genomic_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(.+):(\d+)-(\d+)$&quot;</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">genomic</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">genomic_match</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: no a valid genomic &quot;</span>
                              <span class="s2">&quot;region&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">genomic</span><span class="p">))</span>
        <span class="n">chrom</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">genomic_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>
        <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">genomic_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>
        <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">genomic_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">3</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">chrom</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: invalid chromosome&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">end</span> <span class="o">&lt;</span> <span class="n">start</span><span class="p">:</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">end</span><span class="p">,</span> <span class="n">start</span>

        <span class="n">GenomicRange</span> <span class="o">=</span> <span class="n">namedtuple</span><span class="p">(</span><span class="s2">&quot;GenomicRange&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">])</span>
        <span class="n">args</span><span class="o">.</span><span class="n">genomic</span> <span class="o">=</span> <span class="n">GenomicRange</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>

    <span class="c1"># If MAF, we check what&#39;s required</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">maf</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">extensions</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="s2">&quot;map&quot;</span><span class="p">)</span>
        <span class="n">extensions</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="s2">&quot;maf&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">maf</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">maf</span> <span class="o">&gt;</span> <span class="mf">0.5</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: invalid MAF&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">maf</span><span class="p">))</span>

    <span class="c1"># If completion rate, we check what&#39;s required</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">rate</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">extensions</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="s2">&quot;map&quot;</span><span class="p">)</span>
        <span class="n">extensions</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="s2">&quot;completion_rates&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">rate</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">rate</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: invalid rate&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">rate</span><span class="p">))</span>

    <span class="c1"># If info, we check what&#39;s required</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">info</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">extensions</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="s2">&quot;map&quot;</span><span class="p">)</span>
        <span class="n">extensions</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="s2">&quot;impute2_info&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">info</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">info</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: invalid information &quot;</span>
                              <span class="s2">&quot;value&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">info</span><span class="p">))</span>

    <span class="c1"># Checking the probability threshold</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">prob</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">prob</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: invalid probability &quot;</span>
                          <span class="s2">&quot;threshold&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">prob</span><span class="p">))</span>

    <span class="c1"># Checking the companion files (for impute2)</span>
    <span class="n">f_prefix</span> <span class="o">=</span> <span class="n">get_file_prefix</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">impute2</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">f_extension</span> <span class="ow">in</span> <span class="n">extensions</span><span class="p">:</span>
        <span class="n">fn</span> <span class="o">=</span> <span class="n">f_prefix</span> <span class="o">+</span> <span class="s2">&quot;.&quot;</span> <span class="o">+</span> <span class="n">f_extension</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">fn</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fn</span><span class="p">))</span>

    <span class="c1"># Checking the output format</span>
    <span class="k">for</span> <span class="n">out_format</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">out_format</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">out_format</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;impute2&quot;</span><span class="p">,</span> <span class="s2">&quot;dosage&quot;</span><span class="p">,</span> <span class="s2">&quot;calls&quot;</span><span class="p">,</span> <span class="s2">&quot;bed&quot;</span><span class="p">}:</span>
            <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: invalid output format&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">out_format</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">out_format</span> <span class="o">==</span> <span class="s2">&quot;bed&quot;</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">HAS_PYPLINK</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;missing optional module: pyplink&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">out_format</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;bed&quot;</span><span class="p">,</span> <span class="s2">&quot;calls&quot;</span><span class="p">,</span> <span class="s2">&quot;dosage&quot;</span><span class="p">}:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">f_prefix</span> <span class="o">+</span> <span class="s2">&quot;.sample&quot;</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">GenipeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: sample file missing&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">f_prefix</span> <span class="o">+</span> <span class="s2">&quot;.sample&quot;</span><span class="p">),</span>
                <span class="p">)</span>

    <span class="k">return</span> <span class="kc">True</span></div>


<div class="viewcode-block" id="parse_args"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_extractor.parse_args">[docs]</a><span class="k">def</span> <span class="nf">parse_args</span><span class="p">(</span><span class="n">parser</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Parses the command line options and arguments.</span>

<span class="sd">    Args:</span>
<span class="sd">        parser (argparse.ArgumentParser): the argument parser</span>
<span class="sd">        args (list): the list of arguments (if not taken from ``sys.argv``)</span>

<span class="sd">    Returns:</span>
<span class="sd">        argparse.Namespace: the list of options and arguments</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">        The only check that is done here is by the parser itself. Values are</span>
<span class="sd">        verified later by the :py:func:`check_args` function.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Adding the version option for the main parser</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-v&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--version&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;version&quot;</span><span class="p">,</span>
        <span class="n">version</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(prog)s</span><span class="s2">, part of genipe version </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--debug&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;set the logging level to debug&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># The input files</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Input Files&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--impute2&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;The output from IMPUTE2.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Indexation options</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Indexation Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--index&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;index_only&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Only perform the indexation.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># The output files</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Output Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--out&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PREFIX&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;impute2_extractor&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;The prefix of the output files. [</span><span class="si">%(default)s</span><span class="s2">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--format&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FORMAT&quot;</span><span class="p">,</span>
        <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;impute2&quot;</span><span class="p">],</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;out_format&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;The output format. Can specify either &#39;impute2&#39; for &quot;</span>
             <span class="s2">&quot;probabilities (same as impute2 format, i.e. 3 values per &quot;</span>
             <span class="s2">&quot;sample), &#39;dosage&#39; for dosage values (one value between 0 and 2 &quot;</span>
             <span class="s2">&quot;by sample), &#39;calls&#39; for hard calls, or &#39;bed&#39; for Plink binary &quot;</span>
             <span class="s2">&quot;format (with hard calls). </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--long&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;long_format&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Write the output file in the long format (one line per sample &quot;</span>
             <span class="s2">&quot;per marker). This option is only compatible with the &#39;calls&#39; &quot;</span>
             <span class="s2">&quot;and &#39;dosage&#39; format (option &#39;--format&#39;).&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--prob&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FLOAT&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;The probability threshold used when creating a file in the &quot;</span>
             <span class="s2">&quot;dosage or call format. [</span><span class="si">%(default).1f</span><span class="s2">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># What to extract</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Extraction Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--extract&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;File containing marker names to extract.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--genomic&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;CHR:START-END&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;The range to extract (e.g. 22 1000000 1500000). Can be use in &quot;</span>
             <span class="s2">&quot;combination with &#39;--rate&#39;, &#39;--maf&#39; and &#39;--info&#39;.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--maf&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FLOAT&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Extract markers with a minor allele frequency equal or higher &quot;</span>
             <span class="s2">&quot;than the specified threshold. Can be use in combination with &quot;</span>
             <span class="s2">&quot;&#39;--rate&#39;, &#39;--info&#39; and &#39;--genomic&#39;.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--rate&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FLOAT&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Extract markers with a completion rate equal or higher to the &quot;</span>
             <span class="s2">&quot;specified threshold. Can be use in combination with &#39;--maf&#39;, &quot;</span>
             <span class="s2">&quot;&#39;--info&#39; and &#39;--genomic&#39;.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--info&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FLOAT&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Extract markers with an information equal or higher to the &quot;</span>
             <span class="s2">&quot;specified threshold. Can be use in combination with &#39;--maf&#39;, &quot;</span>
             <span class="s2">&quot;&#39;--rate&#39; and &#39;--genomic&#39;.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span></div>


<span class="c1"># Calling the main, if necessary</span>
<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">genipe 1.4.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../../genipe.html" >genipe</a> &#187;</li> 
      </ul>
    </div>


    <div class="footer" role="contentinfo">
        &#169; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.7.5.
    </div>

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-81009922-1', 'auto');
  ga('send', 'pageview');

</script>


  </body>
</html>